{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Conversion Appyter\n",
    "\n",
    "Here we convert a transcript expression count matrix into a gene expression count matrix leveraging official NCBI Symbols.\n",
    "\n",
    "See: <ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/> for more information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Gene Conversion',\n",
    "    subtitle='Configure your gene conversion',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: Load the input data\n",
    "\n",
    "It's important to note that the **first column** of your matrix should contain the symbols we're converting, and the **first row** the column header. More advanced tabular formats are not yet supported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do DescriptionField(\n",
    "    name='desc',\n",
    "    text='''\n",
    "      <div class=\"alert alert-info mb-0\">\n",
    "      Your input matrix should have the gene identifiers to be\n",
    "      converted in the <b>first column</b> of the matrix, and\n",
    "      at most <b>one row of header</b>. In the case of Excel which\n",
    "      is, in general, not recommended due to its tendency to convert\n",
    "      gene names into dates, only the first sheet will be used.\n",
    "      </div>\n",
    "    ''',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set file = FileField(\n",
    "  name='file',\n",
    "  label='File to convert',\n",
    "  description='Tab-separated matrix to be converted from transcripts to genes',\n",
    "  default='GSE152418.txt.gz',\n",
    "  examples={\n",
    "    'GSE152418.txt.gz': 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE152418&format=file&file=GSE152418%5Fp20047%5FStudy1%5FRawCounts%2Etxt%2Egz',\n",
    "  },\n",
    "  section='primary',\n",
    ") %}\n",
    "\n",
    "{% set file_format = ChoiceField(\n",
    "  name='file_format',\n",
    "  label='File Format',\n",
    "  description='Please select your file format',\n",
    "  default='TSV (.tsv / .txt)',\n",
    "  choices={\n",
    "    'TSV (.tsv / .txt)': \"sep='\\\\t',\",\n",
    "    'GZipped TSV (.tsv.gz / .txt.gz)': \"sep='\\\\t', compression='gzip',\",\n",
    "    'CSV (.csv)': \"sep=',',\",\n",
    "    'GZipped CSV (.csv.gz)': \"compression='gzip',\",\n",
    "    'Excel Sheet 1 (.xls, .xlsx, .xlsm, .xlsb, .odf, .ods, .odt)': \"read_excel\",\n",
    "  },\n",
    "  section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "data = pd.read_csv(\n",
       "    'GSE152418.txt.gz',\n",
       "    index_col=0,\n",
       "    sep='\\t',\n",
       ")\n",
       "data\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>S145_nCOV001_C</th>\n",
       "      <th>S147_nCoV001EUHM-Draw-1</th>\n",
       "      <th>S149_nCoV002EUHM-Draw-2</th>\n",
       "      <th>S150_nCoV003EUHM-Draw-1</th>\n",
       "      <th>S151_nCoV004EUHM-Draw-1</th>\n",
       "      <th>S152_nCoV006EUHM-Draw-1</th>\n",
       "      <th>S153_nCoV007EUHM-Draw-1</th>\n",
       "      <th>S154_nCoV0010EUHM-Draw-1</th>\n",
       "      <th>S155_nCOV021EUHM</th>\n",
       "      <th>S156_nCOV024EUHM-Draw-1</th>\n",
       "      <th>...</th>\n",
       "      <th>S068_272</th>\n",
       "      <th>S069_273</th>\n",
       "      <th>S070_279</th>\n",
       "      <th>S071_280</th>\n",
       "      <th>S181_255</th>\n",
       "      <th>S182_SHXA10</th>\n",
       "      <th>S183_263</th>\n",
       "      <th>S184_SHXA18</th>\n",
       "      <th>S185_266</th>\n",
       "      <th>S186_SHXA14</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSEMBLID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284332</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>40</td>\n",
       "      <td>31</td>\n",
       "      <td>56</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>66</td>\n",
       "      <td>40</td>\n",
       "      <td>24</td>\n",
       "      <td>32</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>62</td>\n",
       "      <td>48</td>\n",
       "      <td>88</td>\n",
       "      <td>38</td>\n",
       "      <td>63</td>\n",
       "      <td>62</td>\n",
       "      <td>83</td>\n",
       "      <td>30</td>\n",
       "      <td>60</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275405</th>\n",
       "      <td>31</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275987</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60683 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 S145_nCOV001_C  S147_nCoV001EUHM-Draw-1  \\\n",
       "ENSEMBLID                                                  \n",
       "ENSG00000223972               0                        0   \n",
       "ENSG00000227232               1                        0   \n",
       "ENSG00000278267               3                        1   \n",
       "ENSG00000243485               2                        0   \n",
       "ENSG00000284332               0                        0   \n",
       "...                         ...                      ...   \n",
       "ENSG00000271254              40                       31   \n",
       "ENSG00000275405              31                       22   \n",
       "ENSG00000275987               0                        0   \n",
       "ENSG00000277475               0                        0   \n",
       "ENSG00000268674               0                        0   \n",
       "\n",
       "                 S149_nCoV002EUHM-Draw-2  S150_nCoV003EUHM-Draw-1  \\\n",
       "ENSEMBLID                                                           \n",
       "ENSG00000223972                        0                        0   \n",
       "ENSG00000227232                        3                       16   \n",
       "ENSG00000278267                        2                        0   \n",
       "ENSG00000243485                        0                        1   \n",
       "ENSG00000284332                        0                        0   \n",
       "...                                  ...                      ...   \n",
       "ENSG00000271254                       56                       42   \n",
       "ENSG00000275405                       13                       19   \n",
       "ENSG00000275987                        0                        0   \n",
       "ENSG00000277475                        0                        0   \n",
       "ENSG00000268674                        0                        0   \n",
       "\n",
       "                 S151_nCoV004EUHM-Draw-1  S152_nCoV006EUHM-Draw-1  \\\n",
       "ENSEMBLID                                                           \n",
       "ENSG00000223972                        0                        0   \n",
       "ENSG00000227232                        1                        2   \n",
       "ENSG00000278267                        3                        0   \n",
       "ENSG00000243485                        0                        0   \n",
       "ENSG00000284332                        0                        0   \n",
       "...                                  ...                      ...   \n",
       "ENSG00000271254                       42                       66   \n",
       "ENSG00000275405                        5                        4   \n",
       "ENSG00000275987                        0                        0   \n",
       "ENSG00000277475                        1                        0   \n",
       "ENSG00000268674                        0                        0   \n",
       "\n",
       "                 S153_nCoV007EUHM-Draw-1  S154_nCoV0010EUHM-Draw-1  \\\n",
       "ENSEMBLID                                                            \n",
       "ENSG00000223972                        0                         0   \n",
       "ENSG00000227232                       18                         6   \n",
       "ENSG00000278267                        0                         1   \n",
       "ENSG00000243485                        0                         0   \n",
       "ENSG00000284332                        0                         0   \n",
       "...                                  ...                       ...   \n",
       "ENSG00000271254                       40                        24   \n",
       "ENSG00000275405                       18                         6   \n",
       "ENSG00000275987                        0                         0   \n",
       "ENSG00000277475                        0                         0   \n",
       "ENSG00000268674                        0                         0   \n",
       "\n",
       "                 S155_nCOV021EUHM  S156_nCOV024EUHM-Draw-1  ...  S068_272  \\\n",
       "ENSEMBLID                                                   ...             \n",
       "ENSG00000223972                 2                        0  ...         0   \n",
       "ENSG00000227232                20                        0  ...         1   \n",
       "ENSG00000278267                 3                        2  ...         2   \n",
       "ENSG00000243485                 0                        0  ...         0   \n",
       "ENSG00000284332                 0                        0  ...         0   \n",
       "...                           ...                      ...  ...       ...   \n",
       "ENSG00000271254                32                       49  ...        62   \n",
       "ENSG00000275405                 6                        4  ...         8   \n",
       "ENSG00000275987                 0                        0  ...         0   \n",
       "ENSG00000277475                 2                        0  ...         0   \n",
       "ENSG00000268674                 0                        0  ...         0   \n",
       "\n",
       "                 S069_273  S070_279  S071_280  S181_255  S182_SHXA10  \\\n",
       "ENSEMBLID                                                              \n",
       "ENSG00000223972         0         0         0         0            0   \n",
       "ENSG00000227232         1         1         2         3            4   \n",
       "ENSG00000278267         7         2         6         3            2   \n",
       "ENSG00000243485         0         0         1         0            0   \n",
       "ENSG00000284332         0         0         0         0            0   \n",
       "...                   ...       ...       ...       ...          ...   \n",
       "ENSG00000271254        48        88        38        63           62   \n",
       "ENSG00000275405         6        14        15         2            2   \n",
       "ENSG00000275987         0         0         0         0            0   \n",
       "ENSG00000277475         0         0         0         1            0   \n",
       "ENSG00000268674         0         0         0         0            0   \n",
       "\n",
       "                 S183_263  S184_SHXA18  S185_266  S186_SHXA14  \n",
       "ENSEMBLID                                                      \n",
       "ENSG00000223972         0            0         0            0  \n",
       "ENSG00000227232         1            0         0            7  \n",
       "ENSG00000278267         6           12         5            3  \n",
       "ENSG00000243485         0            0         0            0  \n",
       "ENSG00000284332         0            0         0            0  \n",
       "...                   ...          ...       ...          ...  \n",
       "ENSG00000271254        83           30        60           77  \n",
       "ENSG00000275405         7            5        15            3  \n",
       "ENSG00000275987         0            0         0            0  \n",
       "ENSG00000277475         0            0         0            0  \n",
       "ENSG00000268674         0            0         0            0  \n",
       "\n",
       "[60683 rows x 34 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "{% if file_format.value == \"read_excel\" %}\n",
    "data = pd.read_excel(\n",
    "    {{ file }},\n",
    "    index_col=0,\n",
    ")\n",
    "{% else %}\n",
    "data = pd.read_csv(\n",
    "    {{ file }},\n",
    "    index_col=0,\n",
    "    {{ file_format }}\n",
    ")\n",
    "{% endif %}\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set organism = ChoiceField(\n",
    "  name='organism',\n",
    "  label='Select the organism',\n",
    "  description='Different organisms have different sets of genes available; all NCBI supported organisms can be supported.',\n",
    "  default='Homo sapiens',\n",
    "  choices={\n",
    "    'Homo sapiens': 'Mammalia/Homo_sapiens',\n",
    "    'Mus musculus': 'Mammalia/Mus_musculus',\n",
    "    'Rattus norvegicus': 'Mammalia/Rattus_norvegicus',\n",
    "    'Pan troglodytes': 'Mammalia/Pan_troglodytes',\n",
    "    'Sus scrofa': 'Mammalia/Sus_scrofa',\n",
    "    'Canis familiaris': 'Mammalia/Canis_familiaris',\n",
    "    'Bos taurus': 'Mammalia/Bos_taurus',\n",
    "  },\n",
    "  section='primary',\n",
    ") %}\n",
    "\n",
    "{% set gene_types = MultiChoiceField(\n",
    "  name='gene_types',\n",
    "  label='Gene types',\n",
    "  description='Types of genes to include in the mapping',\n",
    "  default=['protein-coding'],\n",
    "  choices=[\n",
    "    'protein-coding',\n",
    "    'ncRNA',\n",
    "    'pseudo',\n",
    "    'biological-region',\n",
    "    'unknown',\n",
    "    'other',\n",
    "    'tRNA',\n",
    "    'snoRNA',\n",
    "    'snRNA',\n",
    "    'rRNA',\n",
    "    'scRNA',\n",
    "  ],\n",
    "  section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "ncbi_gene_info = pd.read_csv(\n",
       "    'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz',\n",
       "    compression='gzip', sep='\\t',\n",
       ")\n",
       "ncbi_gene_info = ncbi_gene_info[np.in1d(ncbi_gene_info['type_of_gene'], ['protein-coding'])]\n",
       "ncbi_gene_info\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20201124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20201124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20201124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20201124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9606</td>\n",
       "      <td>12</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>-</td>\n",
       "      <td>AACT|ACT|GIG24|GIG25</td>\n",
       "      <td>MIM:107280|HGNC:HGNC:16|Ensembl:ENSG00000196136</td>\n",
       "      <td>14</td>\n",
       "      <td>14q32.13</td>\n",
       "      <td>serpin family A member 3</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>serpin family A member 3</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1-antichymotrypsin|cell growth-inhibitin...</td>\n",
       "      <td>20201124</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61730</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923209</td>\n",
       "      <td>ND1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 1</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61731</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923210</td>\n",
       "      <td>ND5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 5</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 5</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61732</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923211</td>\n",
       "      <td>ATP8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>ATP synthase F0 subunit 8</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ATP synthase F0 subunit 8</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61733</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923212</td>\n",
       "      <td>ND2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NADH dehydrogenase subunit 2</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61739</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923218</td>\n",
       "      <td>COX1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19853 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #tax_id   GeneID    Symbol LocusTag                   Synonyms  \\\n",
       "0         9606        1      A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1         9606        2       A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "3         9606        9      NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4         9606       10      NAT2        -            AAC2|NAT-2|PNAT   \n",
       "6         9606       12  SERPINA3        -       AACT|ACT|GIG24|GIG25   \n",
       "...        ...      ...       ...      ...                        ...   \n",
       "61730   741158  8923209       ND1        -                          -   \n",
       "61731   741158  8923210       ND5        -                          -   \n",
       "61732   741158  8923211      ATP8        -                          -   \n",
       "61733   741158  8923212       ND2        -                          -   \n",
       "61739   741158  8923218      COX1        -                          -   \n",
       "\n",
       "                                                 dbXrefs chromosome  \\\n",
       "0         MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19   \n",
       "1         MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12   \n",
       "3      MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8   \n",
       "4      MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8   \n",
       "6        MIM:107280|HGNC:HGNC:16|Ensembl:ENSG00000196136         14   \n",
       "...                                                  ...        ...   \n",
       "61730                                                  -         MT   \n",
       "61731                                                  -         MT   \n",
       "61732                                                  -         MT   \n",
       "61733                                                  -         MT   \n",
       "61739                                                  -         MT   \n",
       "\n",
       "      map_location                     description    type_of_gene  \\\n",
       "0         19q13.43          alpha-1-B glycoprotein  protein-coding   \n",
       "1         12p13.31           alpha-2-macroglobulin  protein-coding   \n",
       "3             8p22           N-acetyltransferase 1  protein-coding   \n",
       "4             8p22           N-acetyltransferase 2  protein-coding   \n",
       "6         14q32.13        serpin family A member 3  protein-coding   \n",
       "...            ...                             ...             ...   \n",
       "61730            -    NADH dehydrogenase subunit 1  protein-coding   \n",
       "61731            -    NADH dehydrogenase subunit 5  protein-coding   \n",
       "61732            -       ATP synthase F0 subunit 8  protein-coding   \n",
       "61733            -    NADH dehydrogenase subunit 2  protein-coding   \n",
       "61739            -  cytochrome c oxidase subunit I  protein-coding   \n",
       "\n",
       "      Symbol_from_nomenclature_authority  \\\n",
       "0                                   A1BG   \n",
       "1                                    A2M   \n",
       "3                                   NAT1   \n",
       "4                                   NAT2   \n",
       "6                               SERPINA3   \n",
       "...                                  ...   \n",
       "61730                                  -   \n",
       "61731                                  -   \n",
       "61732                                  -   \n",
       "61733                                  -   \n",
       "61739                                  -   \n",
       "\n",
       "      Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                    alpha-1-B glycoprotein                   O   \n",
       "1                     alpha-2-macroglobulin                   O   \n",
       "3                     N-acetyltransferase 1                   O   \n",
       "4                     N-acetyltransferase 2                   O   \n",
       "6                  serpin family A member 3                   O   \n",
       "...                                     ...                 ...   \n",
       "61730                                     -                   -   \n",
       "61731                                     -                   -   \n",
       "61732                                     -                   -   \n",
       "61733                                     -                   -   \n",
       "61739                                     -                   -   \n",
       "\n",
       "                                      Other_designations  Modification_date  \\\n",
       "0      alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20201124   \n",
       "1      alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20201124   \n",
       "3      arylamine N-acetyltransferase 1|N-acetyltransf...           20201124   \n",
       "4      arylamine N-acetyltransferase 2|N-acetyltransf...           20201124   \n",
       "6      alpha-1-antichymotrypsin|cell growth-inhibitin...           20201124   \n",
       "...                                                  ...                ...   \n",
       "61730                       NADH dehydrogenase subunit 1           20200909   \n",
       "61731                       NADH dehydrogenase subunit 5           20200909   \n",
       "61732                          ATP synthase F0 subunit 8           20200909   \n",
       "61733                       NADH dehydrogenase subunit 2           20200909   \n",
       "61739                     cytochrome c oxidase subunit I           20200909   \n",
       "\n",
       "      Feature_type  \n",
       "0                -  \n",
       "1                -  \n",
       "3                -  \n",
       "4                -  \n",
       "6                -  \n",
       "...            ...  \n",
       "61730            -  \n",
       "61731            -  \n",
       "61732            -  \n",
       "61733            -  \n",
       "61739            -  \n",
       "\n",
       "[19853 rows x 16 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "ncbi_gene_info = pd.read_csv(\n",
    "    'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/{{ organism }}.gene_info.gz',\n",
    "    compression='gzip', sep='\\t',\n",
    ")\n",
    "ncbi_gene_info = ncbi_gene_info[np.in1d(ncbi_gene_info['type_of_gene'], {{ gene_types }})]\n",
    "ncbi_gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1. Prepare NCBI Gene Synonym lookup table\n",
    "\n",
    "To perform this conversion, we'll prepare a lookup table with the ncbi gene info table. Then we'll use that lookup for the conversion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HGNC:1471                                                                CAMLG\n",
       "lc3 synthase                                                            B3GNT5\n",
       "ENSG00000170703                                                          TTLL6\n",
       "HGNC:4388                                                                 GNAL\n",
       "HNP3                                                                     DEFA3\n",
       "                                                                        ...   \n",
       "pyruvate dehydrogenase, E1 beta polypeptide                               PDHB\n",
       "pyruvate dehydrogenase, lipoamide, kinase isozyme 3, mitochondrial        PDK3\n",
       "Ensembl:ENSG00000171564                                                    FGB\n",
       "HGNC:HGNC:23536                                                       NAALADL1\n",
       "MIM:607969                                                                CLK4\n",
       "Length: 281717, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a lookup dictionary using the ncbi table\n",
    "\n",
    "def maybe_split(record):\n",
    "  ''' NCBI Stores Nulls as '-' and lists '|' delimited\n",
    "  '''\n",
    "  if record in {'', '-'}:\n",
    "    return set()\n",
    "  return set(record.split('|'))\n",
    "\n",
    "def supplement_dbXref_prefix_omitted(ids):\n",
    "  ''' NCBI Stores external IDS with Foreign:ID while most datasets just use the ID\n",
    "  '''\n",
    "  for id in ids:\n",
    "    # add original id\n",
    "    yield id\n",
    "    # also add id *without* prefix\n",
    "    if ':' in id:\n",
    "      yield id.split(':', maxsplit=1)[1]\n",
    "\n",
    "synonym, symbol = zip(*{\n",
    "  (alternate_symbol, gene_info['Symbol'])\n",
    "  for _, gene_info in ncbi_gene_info.iterrows()\n",
    "  for alternate_symbol in set.union(\n",
    "    maybe_split(gene_info['Symbol']),\n",
    "    maybe_split(gene_info['Symbol_from_nomenclature_authority']),\n",
    "    maybe_split(str(gene_info['GeneID'])),\n",
    "    maybe_split(gene_info['Synonyms']),\n",
    "    maybe_split(gene_info['Other_designations']),\n",
    "    maybe_split(gene_info['LocusTag']),\n",
    "    set(supplement_dbXref_prefix_omitted(maybe_split(gene_info['dbXrefs']))),\n",
    "  )\n",
    "})\n",
    "ncbi_lookup = pd.Series(symbol, index=synonym)\n",
    "ncbi_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "epididymis secretory sperm binding protein    387\n",
       "seven transmembrane helix receptor             49\n",
       "flavoprotein-linked monooxygenase              16\n",
       "microsomal monooxygenase                       15\n",
       "H4C4                                           14\n",
       "                                             ... \n",
       "VPS4-1                                          1\n",
       "HGNC:HGNC:10729                                 1\n",
       "HGNC:6816                                       1\n",
       "HGNC:9176                                       1\n",
       "C16orf41                                        1\n",
       "Length: 274553, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# turns out some synonyms are ambiguous\n",
    "index_values = ncbi_lookup.index.value_counts()\n",
    "index_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HGNC:1471                                                                CAMLG\n",
       "lc3 synthase                                                            B3GNT5\n",
       "ENSG00000170703                                                          TTLL6\n",
       "HGNC:4388                                                                 GNAL\n",
       "HNP3                                                                     DEFA3\n",
       "                                                                        ...   \n",
       "pyruvate dehydrogenase, E1 beta polypeptide                               PDHB\n",
       "pyruvate dehydrogenase, lipoamide, kinase isozyme 3, mitochondrial        PDK3\n",
       "Ensembl:ENSG00000171564                                                    FGB\n",
       "HGNC:HGNC:23536                                                       NAALADL1\n",
       "MIM:607969                                                                CLK4\n",
       "Length: 269494, dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for our sanity we'll drop these entirely, in the future we could consider a better way to take care of these\n",
    "ncbi_lookup_disambiguated = ncbi_lookup.drop(index_values[index_values > 1].index)\n",
    "ncbi_lookup_disambiguated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Use Disambiguated NCBI lookup table to map our data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% set versioned_ensembl_id = BoolField(\n",
    "  name='versioned_ensembl_id',\n",
    "  label='Do you have versioned ENSEMBL IDs?',\n",
    "  description='ENSEMBL Ids may be postfixed with the version e.g `ENSG00000227232.1`. If your ENSEMBL ids are versioned, you can check this to prune the versions for a successful conversion.',\n",
    "  default=False,\n",
    "  section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "```python\n",
       "# Mapping between current transcripts to the official NCBI Gene symbols, dropping anything that doesn't map\n",
       "ncbi_lookup_disambiguated_dict = ncbi_lookup_disambiguated.to_dict()\n",
       "transcript_genes = pd.Series(\n",
       "    data.index.map(ncbi_lookup_disambiguated_dict.get),\n",
       "    index=data.index,\n",
       ").dropna()\n",
       "transcript_genes\n",
       "```"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "ENSEMBLID\n",
       "ENSG00000186092           OR4F5\n",
       "ENSG00000284733          OR4F29\n",
       "ENSG00000284662          OR4F16\n",
       "ENSG00000187634          SAMD11\n",
       "ENSG00000188976           NOC2L\n",
       "                       ...     \n",
       "ENSG00000277196    LOC102724788\n",
       "ENSG00000277630    LOC100288966\n",
       "ENSG00000278633    LOC102724151\n",
       "ENSG00000276345    LOC107987373\n",
       "ENSG00000271254    LOC102724250\n",
       "Name: ENSEMBLID, Length: 19286, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "# Mapping between current transcripts to the official NCBI Gene symbols, dropping anything that doesn't map\n",
    "ncbi_lookup_disambiguated_dict = ncbi_lookup_disambiguated.to_dict()\n",
    "transcript_genes = pd.Series(\n",
    "{%- if versioned_ensembl_id.value %}\n",
    "    data.index.map(lambda i: ncbi_lookup_disambiguated_dict.get(i.split('.')[0])),\n",
    "{%- else %}\n",
    "    data.index.map(ncbi_lookup_disambiguated_dict.get),\n",
    "{%- endif %}\n",
    "    index=data.index,\n",
    ").dropna()\n",
    "transcript_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSEMBLID\n",
       "ENSG00000223972      0.270945\n",
       "ENSG00000227232     51.871658\n",
       "ENSG00000278267      8.074866\n",
       "ENSG00000243485      0.167558\n",
       "ENSG00000284332      0.000000\n",
       "                      ...    \n",
       "ENSG00000271254    573.469697\n",
       "ENSG00000275405    118.106952\n",
       "ENSG00000275987      0.117647\n",
       "ENSG00000277475      0.528520\n",
       "ENSG00000268674      0.000000\n",
       "Length: 60683, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute variance of each transcript\n",
    "transcript_variance = data.var(axis=1)\n",
    "transcript_variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variance</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSEMBLID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>0.311052</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284733</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000284662</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187634</th>\n",
       "      <td>3.910873</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>97646.467914</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277196</th>\n",
       "      <td>1.476827</td>\n",
       "      <td>LOC102724788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277630</th>\n",
       "      <td>0.029412</td>\n",
       "      <td>LOC100288966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278633</th>\n",
       "      <td>0.264706</td>\n",
       "      <td>LOC102724151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000276345</th>\n",
       "      <td>625.450980</td>\n",
       "      <td>LOC107987373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>573.469697</td>\n",
       "      <td>LOC102724250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19286 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     variance          gene\n",
       "ENSEMBLID                                  \n",
       "ENSG00000186092      0.311052         OR4F5\n",
       "ENSG00000284733      0.000000        OR4F29\n",
       "ENSG00000284662      0.000000        OR4F16\n",
       "ENSG00000187634      3.910873        SAMD11\n",
       "ENSG00000188976  97646.467914         NOC2L\n",
       "...                       ...           ...\n",
       "ENSG00000277196      1.476827  LOC102724788\n",
       "ENSG00000277630      0.029412  LOC100288966\n",
       "ENSG00000278633      0.264706  LOC102724151\n",
       "ENSG00000276345    625.450980  LOC107987373\n",
       "ENSG00000271254    573.469697  LOC102724250\n",
       "\n",
       "[19286 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the transcript_genes & transcript_variance\n",
    "transcript_gene_variance = pd.merge(\n",
    "    left=transcript_variance.to_frame('variance'), left_index=True,\n",
    "    right=transcript_genes.to_frame('gene'), right_index=True,\n",
    ")\n",
    "transcript_gene_variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "A1BG       ENSG00000121410\n",
       "A1CF       ENSG00000148584\n",
       "A2M        ENSG00000175899\n",
       "A2ML1      ENSG00000166535\n",
       "A3GALT2    ENSG00000184389\n",
       "                ...       \n",
       "ZYG11A     ENSG00000203995\n",
       "ZYG11B     ENSG00000162378\n",
       "ZYX        ENSG00000159840\n",
       "ZZEF1      ENSG00000074755\n",
       "ZZZ3       ENSG00000036549\n",
       "Name: variance, Length: 19233, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for all transcripts that map to multiple genes, take the gene, transcript mapping with the highest variance\n",
    "transcript_gene_mapping = transcript_gene_variance.groupby('gene')['variance'].idxmax()\n",
    "transcript_gene_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# actually perform the mapping\n",
    "data_genes = pd.merge(\n",
    "    left=transcript_gene_mapping.to_frame('transcript'), left_on='transcript',\n",
    "    right=data, right_index=True,\n",
    ").drop('transcript', axis=1)\n",
    "data_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "## Step 3. Export results\n",
    "\n",
    "Your results are available for download [here](./converted_{{ file }})."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Converted Matrix of shape {data.shape} => {data_genes.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "{% if file_format.value == \"excel\" %}\n",
    "data_genes.to_excel(\n",
    "    'converted_' + {{ file }},\n",
    ")\n",
    "{% else %}\n",
    "data_genes.to_csv(\n",
    "    'converted_' + {{ file }},\n",
    "    {{ file_format }}\n",
    ")\n",
    "{% endif %}\n",
    "data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
