{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02298265",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fd169e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "# Gene and Protien Expression across Human Cells and Tissues\n",
    "The Appyter takes a human gene symbol as the input to produce box plots that display its expression across human cell types and tissue at the mRNA and protein levels. If the gene is not contained in one of the datasets, a plot will not be produced for that resource."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c870541",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Gene and Protein Expression across Cells and Tissues',\n",
    "    img ='gene-expr.png'\n",
    ") %}\n",
    "\n",
    "{% do DescriptionField(\n",
    "    name='data_file_description',\n",
    "    text='''\n",
    "    This appyter takes a gene symbol as the input and displays its expression across cells and tissues \n",
    "    utilizing a variety of processed datasets from healthy tissues.\n",
    "    If the gene is not contained in one of the datasets, a plot will not be produced for that resource.''',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "%%appyter code_exec\n",
    "{% set user_input = TabField(\n",
    "    name = 'species_input',\n",
    "    label = 'Select species and datasets to display',\n",
    "    default = 'Human',\n",
    "    description = 'Select the desired species and the available datasets',\n",
    "    section = 'primary',\n",
    "    choices = {\n",
    "        'Human': [\n",
    "            AutocompleteField(\n",
    "                name = 'human_gene',\n",
    "                label = 'Gene symbol of interest',\n",
    "                default = 'A2M',\n",
    "                description = 'Enter the gene symbol of interest (human)',\n",
    "                file_path = 'https://appyters.maayanlab.cloud/storage/Gene_Expression_by_Tissue/genes.json',\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='gtex_gene',\n",
    "                label='Include GTEX - gene?',\n",
    "                default = True\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='archs',\n",
    "                label='Include ARCHS4',\n",
    "                default = True\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='ts',\n",
    "                label='Include Tabula Sapiens?',\n",
    "                default = True,\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='ccle',\n",
    "                label='Include CCLE - Transcriptomics?',\n",
    "                default = True,\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='use_hpm',\n",
    "                label='Include HPM?',\n",
    "                default = True,\n",
    "            ),\n",
    "\n",
    "            BoolField(\n",
    "                name='use_hpa',\n",
    "                label='Include HPA?',\n",
    "                default = True,\n",
    "            ),\n",
    "\n",
    "            BoolField(\n",
    "                name='gtex_prot',\n",
    "                label='Include GTEx - Proteomics?',\n",
    "                default = True,\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='ccle-prot',\n",
    "                label='Include CCLE - Proteomics?',\n",
    "                default = True,\n",
    "            ),\n",
    "            \n",
    "        ],\n",
    "        'Mouse': [\n",
    "            AutocompleteField(\n",
    "                name = 'mouse_gene',\n",
    "                label = 'Gene symbol of interest',\n",
    "                default = 'Mthfr',\n",
    "                description = 'Enter the gene symbol of interest (mouse).',\n",
    "                file_path = 'https://appyters.maayanlab.cloud/storage/Gene_Expression_by_Tissue/mm_genes.json',\n",
    "            ),\n",
    "            \n",
    "            BoolField(\n",
    "                name='archs_mm',\n",
    "                label='Include ARCHS4?',\n",
    "                default = True,\n",
    "            ),\n",
    "            \n",
    "            BoolField(\n",
    "                name='tm',\n",
    "                label='Include Tabula Muris?',\n",
    "                default = True,\n",
    "            ),\n",
    "\n",
    "        ]\n",
    "    }\n",
    ")%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aec0415",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import HTML, display, Markdown\n",
    "import s3fs\n",
    "import h5py\n",
    "\n",
    "s3 = s3fs.S3FileSystem(anon=True, client_kwargs={'endpoint_url': 'https://appyters.maayanlab.cloud/storage'})\n",
    "def get_gene_stats(gene, f, source):\n",
    "    if gene not in f[source].keys():\n",
    "        return pd.DataFrame()\n",
    "    if source in ['gtex','ts', 'tm', 'archs4']:\n",
    "        data = f[source][gene]['stats']\n",
    "        cols = np.array(f[source]['columns']).astype('str')\n",
    "        rows = np.array(f[source]['rows']).astype('str')\n",
    "        return pd.DataFrame(data, index=rows, columns=cols)\n",
    "    if source in ['gtex-prot', 'hpm']:\n",
    "        data = []\n",
    "        cols = np.array(f[source]['columns']).astype('str')\n",
    "        for tissue in f[source][gene].keys():\n",
    "            vals = pd.DataFrame(data=np.array(f[source][gene][tissue]), columns=cols)\n",
    "            vals['tissue'] = [tissue] * len(vals)\n",
    "            data.append(vals)\n",
    "        return pd.concat(data)\n",
    "    if source in ['ccle', 'ccle-prot']:\n",
    "        if source == 'ccle':\n",
    "            cols = np.array(f[source]['columns']).astype('str')\n",
    "            vals = np.array(f[source][gene]['stats'])[0]\n",
    "        else:\n",
    "            cols = np.array(f[source][gene]['columns']).astype('str')\n",
    "            vals = np.array(f[source][gene]['stats'])\n",
    "        return pd.Series(data=vals, index=cols)\n",
    "    if source == 'hpa':\n",
    "        cols = np.array(f[source]['columns']).astype('str')\n",
    "        return pd.DataFrame(data=np.array(f[source][gene]['stats']).astype(str), columns=cols)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a542ba08",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "gene = {{ user_input.value[0] }}\n",
    "{% set species = user_input[\"args\"][\"value\"] %}\n",
    "species = {{species|jsonify}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bee96b04",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "## Load in normal gene expression\n",
    "Utilize processed datasets containing gene expression by cell type and tissue from [GTEx](https://gtexportal.org/home/) [1], \n",
    "[ARCHS4](https://maayanlab.cloud/archs4/) [2], and the [Tabula Sapiens](https://tabula-sapiens-portal.ds.czbiohub.org/) [3] and by cell line from [CCLE](https://sites.broadinstitute.org/ccle/) [8].\n",
    "{% endif %}\n",
    "{% if species == 'Mouse' %}\n",
    "Utilize processed datasets containing gene expression by cell type and tissue from \n",
    "[ARCHS4](https://maayanlab.cloud/archs4/) [2] and the [Tabula Muris](https://tabula-muris.ds.czbiohub.org/) [7].\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20014a72",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "s3f = s3.open(\"Gene_Expression_by_Tissue/RNA-Seq-GTEx-ARCHS4-TS-CCLE.h5\")\n",
    "f = h5py.File(s3f, 'r')\n",
    "{% endif %}\n",
    "{% if species == 'Mouse' %}\n",
    "s3f = s3.open(\"Gene_Expression_by_Tissue/mm_RNA-Seq-ARCHS4-TM.h5\")\n",
    "f = h5py.File(s3f, 'r')\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb7f97f-a4ce-4556-acd1-480a65e44fad",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "    {% set gtex_gene = user_input.value[1].value %}\n",
    "    {% set archs = user_input.value[2].value %}\n",
    "    {% set ts = user_input.value[3].value %}\n",
    "    {% set ccle = user_input.value[4].value %}\n",
    "\n",
    "    {% set use_hpm = user_input.value[5].value %}\n",
    "    {% set use_hpa = user_input.value[6].value %}\n",
    "    {% set gtex_prot = user_input.value[7].value %}\n",
    "    {% set ccle_prot = user_input.value[8].value %}\n",
    "{% endif %}\n",
    "{% if species == 'Mouse' %}\n",
    "    {% set archs_mm = user_input.value[1].value %}\n",
    "    {% set tm = user_input.value[2].value %}\n",
    "{% endif %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57340c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "{% if gtex_gene %}\n",
    "df_bg_stats_gtex = get_gene_stats(gene, f, 'gtex')\n",
    "{% endif %}\n",
    "{% if archs %}\n",
    "df_bg_stats_archs4 = get_gene_stats(gene, f, 'archs4')\n",
    "{% endif %}\n",
    "{% if ts %}\n",
    "ts_stats = get_gene_stats(gene, f, 'ts')\n",
    "{% endif %}\n",
    "{% if ccle %}\n",
    "ccle_stats = get_gene_stats(gene, f, 'ccle')\n",
    "{% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8df62db-6b4f-4b72-9da3-d87d892a2424",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% if species == 'Mouse' %}\n",
    "{% if archs_mm %}\n",
    "df_bg_stats_tm = get_gene_stats(gene, f, 'tm')\n",
    "{% endif %}\n",
    "{% if tm %}\n",
    "df_bg_stats_archs4_mm = get_gene_stats(gene, f, 'archs4')\n",
    "{% endif %}\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947e2b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "f.close()\n",
    "s3f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8dadf42",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "# Show available genes in each dataset\n",
    "\n",
    "\n",
    "data = {'Selected Gene': [gene]}\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "{% if gtex_gene %}\n",
    "data['in GTEx - Gene'] = [not df_bg_stats_gtex.empty]\n",
    "{% endif %}\n",
    "{% if archs %}\n",
    "data['in ARCHS4'] = [not df_bg_stats_archs4.empty]\n",
    "{% endif %}\n",
    "{% if ts %}\n",
    "data['in TS'] = [not ts_stats.empty]\n",
    "{% endif %}\n",
    "{% if ccle %}\n",
    "data['in CCLE'] = [not ccle_stats.empty]\n",
    "{% endif %}\n",
    "{% endif %}\n",
    "\n",
    "{% if species == 'Mouse' %}\n",
    "{% if archs_mm %}\n",
    "data['in ARCHS4'] = [not df_bg_stats_archs4_mm.empty]\n",
    "{% endif %}\n",
    "{% if tm %}\n",
    "data['in Tabula Muris'] = [not df_bg_stats_tm.empty]\n",
    "{% endif %}\n",
    "{% endif %}\n",
    "\n",
    "\n",
    "available = pd.DataFrame(data).T\n",
    "display(HTML(available.to_html(notebook=True, escape=False)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520b08ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "## Load plots based on the selected gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "987d0278",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "c = gene\n",
    "{% if species == 'Human' %}\n",
    "{% if gtex_gene %}\n",
    "if not df_bg_stats_gtex.empty:\n",
    "    df_bg_stats_gtex.sort_values('mean', axis=1, inplace=True)\n",
    "    display(Markdown(\"## RNA-Seq Expression Levels\"))\n",
    "    IQR = df_bg_stats_gtex.loc[('75%')]-df_bg_stats_gtex.loc[('25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            df_bg_stats_gtex.loc[('min')],\n",
    "            df_bg_stats_gtex.loc[('25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=df_bg_stats_gtex.loc[('25%')],\n",
    "        median=df_bg_stats_gtex.loc[('50%')],\n",
    "        q3=df_bg_stats_gtex.loc[('75%')],\n",
    "        upperfence=np.minimum(\n",
    "            df_bg_stats_gtex.loc[('max')],\n",
    "            df_bg_stats_gtex.loc[('75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=df_bg_stats_gtex.loc[('mean')],\n",
    "        sd=df_bg_stats_gtex.loc[('std')],\n",
    "        y=df_bg_stats_gtex.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+ \" (RNA-seq) GTEx\", height=1200)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if archs %}\n",
    "if not df_bg_stats_archs4.empty:\n",
    "    df_bg_stats_archs4.sort_values('mean', axis=1, inplace=True)\n",
    "    IQR = df_bg_stats_archs4.loc[('75%')]-df_bg_stats_archs4.loc[('25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            df_bg_stats_archs4.loc[('min')],\n",
    "            df_bg_stats_archs4.loc[('25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=df_bg_stats_archs4.loc[('25%')],\n",
    "        median=df_bg_stats_archs4.loc[('50%')],\n",
    "        q3=df_bg_stats_archs4.loc[('75%')],\n",
    "        upperfence=np.minimum(\n",
    "            df_bg_stats_archs4.loc[('max')],\n",
    "            df_bg_stats_archs4.loc[('75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=df_bg_stats_archs4.loc[('mean')],\n",
    "        sd=df_bg_stats_archs4.loc[('std')],\n",
    "        y=df_bg_stats_archs4.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+ \" (RNA-seq) ARCHS4\", height=8000)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if ts %}\n",
    "if not ts_stats.empty:\n",
    "    ts_stats.sort_values('mean', axis=1, inplace=True)\n",
    "    IQR = ts_stats.loc[('75%')]-ts_stats.loc[('25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            ts_stats.loc[('min')],\n",
    "            ts_stats.loc[('25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=ts_stats.loc[('25%')],\n",
    "        median=ts_stats.loc[('50%')],\n",
    "        q3=ts_stats.loc[('75%')],\n",
    "        upperfence=np.minimum(\n",
    "            ts_stats.loc[('max')],\n",
    "            ts_stats.loc[('75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=ts_stats.loc[('mean')],\n",
    "        sd=ts_stats.loc[('std')],\n",
    "        y=ts_stats.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+ \" (RNA-seq) Tabula Sapiens\", height=8000)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if ccle %}\n",
    "if not ccle_stats.empty:\n",
    "    ccle_stats = ccle_stats.sort_values()[-100:]\n",
    "    fig = px.scatter(y=list(ccle_stats.index.values), x=(ccle_stats.values), height=4000)\n",
    "    fig.update_layout(title = c+ \" CCLE Transcriptomics (Top 100 Cell Lines)\", yaxis=dict(dtick = 1, tickmode='linear', tickfont = dict(size = 12)))\n",
    "    fig.update_yaxes(title=None, range=[-.5,len(ccle_stats)+.5], tickvals=ccle_stats.index)\n",
    "    fig.update_xaxes(title=None)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "{% endif %}\n",
    "\n",
    "{% if species == 'Mouse' %}\n",
    "{% if archs_mm %}\n",
    "if not df_bg_stats_archs4_mm.empty:\n",
    "    df_bg_stats_archs4_mm.sort_values('mean', axis=1, inplace=True)\n",
    "    IQR = df_bg_stats_archs4_mm.loc[('75%')]-df_bg_stats_archs4_mm.loc[('25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            df_bg_stats_archs4_mm.loc[('min')],\n",
    "            df_bg_stats_archs4_mm.loc[('25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=df_bg_stats_archs4_mm.loc[('25%')],\n",
    "        median=df_bg_stats_archs4_mm.loc[('50%')],\n",
    "        q3=df_bg_stats_archs4_mm.loc[('75%')],\n",
    "        upperfence=np.minimum(\n",
    "            df_bg_stats_archs4_mm.loc[('max')],\n",
    "            df_bg_stats_archs4_mm.loc[('75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=df_bg_stats_archs4_mm.loc[('mean')],\n",
    "        sd=df_bg_stats_archs4_mm.loc[('std')],\n",
    "        y=df_bg_stats_archs4_mm.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+ \" (RNA-seq) ARCHS4\", height=8000)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if tm %}\n",
    "if not df_bg_stats_tm.empty:\n",
    "    df_bg_stats_tm.sort_values('mean', axis=1, inplace=True)\n",
    "    IQR = df_bg_stats_tm.loc[('75%')]-df_bg_stats_tm.loc[('25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            df_bg_stats_tm.loc[('min')],\n",
    "            df_bg_stats_tm.loc[('25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=df_bg_stats_tm.loc[('25%')],\n",
    "        median=df_bg_stats_tm.loc[('50%')],\n",
    "        q3=df_bg_stats_tm.loc[('75%')],\n",
    "        upperfence=np.minimum(\n",
    "            df_bg_stats_tm.loc[('max')],\n",
    "            df_bg_stats_tm.loc[('75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=df_bg_stats_tm.loc[('mean')],\n",
    "        sd=df_bg_stats_tm.loc[('std')],\n",
    "        y=df_bg_stats_tm.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+ \" (RNA-seq) Tabula Muris\", height=8000)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "{% endif %}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89086abf",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "{% if species == 'Human' %}\n",
    "## Proteomics Expression Levels\n",
    "\n",
    "Proteomics data were obtained from the [Human Protein Atlas](https://www.proteinatlas.org/about/download) (HPA) [4] \n",
    "with IHC-based expression profiling, the [Human Proteome Map](https://www.humanproteomemap.org/download.php) (HPM) [5] \n",
    "with MS-based expression quantification, the [GTEx proteome project](https://doi.org/10.1016/j.cell.2020.08.036) [6] using TMT MS, and the [CCLE](https://sites.broadinstitute.org/ccle/). \n",
    "These datasets contain protein expression levels detected in normal tissues and cell types. The gene may not be present \n",
    "in the data from each project (see table for which proteomics data are present/absent). Plots show expression levels (HPA), \n",
    "average spectral counts (HPM), or a log-transformed relative abundance (GTEx) by tissue/cell-type for each gene candidate \n",
    "(excluding expression levels from the HPA where the [reliability score](https://www.proteinatlas.org/about/assays+annotation) was uncertain). \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d29c79e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if species == 'Human' %}\n",
    "s3f = s3.open(\"Gene_Expression_by_Tissue/Proteomics-HPM-HPA-GTEx-CCLE.h5\")\n",
    "f = h5py.File(s3f, 'r')\n",
    "\n",
    "{% if use_hpm %}\n",
    "hpm = get_gene_stats(gene, f, 'hpm')\n",
    "{% endif %}\n",
    "{% if use_hpa %}\n",
    "hpa = get_gene_stats(gene, f, 'hpa')\n",
    "{% endif %}\n",
    "{% if gtex_prot %}\n",
    "gtexp = get_gene_stats(gene, f, 'gtex-prot')\n",
    "{% endif %}\n",
    "{% if ccle_prot %}\n",
    "cclep = get_gene_stats(gene, f, 'ccle-prot')\n",
    "{% endif %}\n",
    "\n",
    "f.close()\n",
    "s3f.close()\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77f418d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% if \"Human\" == species %}\n",
    "# Show available genes in each dataset\n",
    "\n",
    "data = {'Selected Gene': [gene]}\n",
    "\n",
    "{% if use_hpm %}\n",
    "data['in HPM'] = [not hpm.empty]\n",
    "{% endif %}\n",
    "\n",
    "{% if use_hpa %}\n",
    "data['in HPA'] =[not hpa.empty]\n",
    "{% endif %}\n",
    "\n",
    "{% if gtex_prot %}\n",
    "data['in GTEx - Proteomics'] = [not gtexp.empty]\n",
    "{% endif %}\n",
    "\n",
    "{% if ccle_prot %}\n",
    "data['in CCLE - Proteomics'] = [not cclep.empty]\n",
    "{% endif %}\n",
    "\n",
    "available = pd.DataFrame(data).T                  \n",
    "                          \n",
    "display(HTML(available.to_html(notebook=True, escape=False)))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d5b9c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% if \"Human\" == species %}\n",
    "{% if gtex_prot %}\n",
    "if not gtexp.empty:\n",
    "    gtexp['tissue_specificity'] = gtexp.tissue_specificity.fillna('NA')\n",
    "    d = gtexp\n",
    "    fig = px.strip(d, y=\"tissue\", x=\"value\",  \n",
    "                   orientation='h',\n",
    "                   stripmode=\"overlay\",\n",
    "                   hover_data=[\"tissue_specificity\"],\n",
    "                   height=30*d['tissue'].nunique())\n",
    "    fig.add_trace(go.Box(x=d['value'],\n",
    "                         y=d['tissue'],\n",
    "                         orientation='h',\n",
    "                         marker=dict(color='#636EFA'),\n",
    "                         name=\"n > 1\"))\n",
    "    fig.update_layout(title=c + \" (GTEx Proteomics)\",\n",
    "                      autosize=True,\n",
    "                      showlegend=False)\n",
    "    fig.update_xaxes(title=\"log2(relative abundance)\")\n",
    "    fig.update_yaxes(title=None)\n",
    "    fig.show()\n",
    "{% endif %}   \n",
    "\n",
    "{% if use_hpm %}\n",
    "if not hpm.empty:\n",
    "    fig = px.scatter(hpm, \n",
    "                     y=\"tissue\", x=\"value\", \n",
    "                     height=20*hpm.shape[0])\n",
    "    fig.update_layout(title=c + \" (HPM)\", \n",
    "                      autosize=True)\n",
    "    fig.update_xaxes(title=\"Average Spectral Counts\")\n",
    "    fig.update_yaxes(title=None)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if use_hpa %}\n",
    "if not hpa.empty:\n",
    "    hpa.Tissue = hpa[\"Tissue\"] + \", \" + hpa[\"Cell.type\"]\n",
    "    hpa = hpa[hpa['Reliability'] != \"Uncertain\"] \n",
    "    fig = px.scatter(hpa, \n",
    "                     y=\"Tissue\", x=\"Level\", \n",
    "                     category_orders={\"Level\": [\"Not detected\", \"Low\", \"Medium\", \"High\"]}, \n",
    "                     hover_data=[\"Reliability\"],  \n",
    "                     hover_name=\"Tissue\",\n",
    "                     height=20*hpa.shape[0])\n",
    "    fig.update_layout(title=c + \" (HPA)\", \n",
    "                      showlegend=False, \n",
    "                      autosize=True, \n",
    "                      xaxis={'tickmode':'array', \n",
    "                             'tickvals':[0, 1, 2, 3], \n",
    "                             'ticktext':[\"Not detected\", \"Low\", \"Medium\", \"High\"]})\n",
    "    fig.update_xaxes(title=\"Tissue Expression Level\")\n",
    "    fig.update_yaxes(title=None)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% if ccle_prot %}\n",
    "if not cclep.empty:\n",
    "    cclep.sort_values(inplace=True)\n",
    "    fig = px.scatter(cclep, \n",
    "                     y=cclep.index, x=cclep.values, \n",
    "                     height=20*cclep.shape[0])\n",
    "    fig.update_layout(title=c + \" (CCLE Proteomics)\", \n",
    "                      showlegend=False, \n",
    "                      autosize=True)\n",
    "    fig.update_xaxes(title=\"Cell Line Expression Level\")\n",
    "    fig.update_yaxes(title=None, range=[-.5,len(cclep)+.5], tickvals= cclep.index)\n",
    "    fig.show()\n",
    "{% endif %}\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5928b4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "# References\n",
    "\n",
    "[1] Lonsdale, John, et al. \"The genotype-tissue expression (GTEx) project.\" Nature genetics 45.6 (2013): 580-585. https://doi.org/10.1038/ng.265\n",
    "        \n",
    "[2] Lachmann A, Torre D, Keenan AB, Jagodnik KM, Lee HJ, Wang L, Silverstein MC, Ma'ayan A. Massive mining of publicly available RNA-seq data from human and mouse. Nature Communications 9. Article number: 1366 (2018), https://doi.org/10.1038/s41467-018-03751-6\n",
    "\n",
    "[3] Tabula Sapiens Consortium* et al. “The Tabula Sapiens: A multiple-organ, single-cell transcriptomic atlas of humans.” Science (New York, N.Y.) vol. 376,6594 (2022). https://doi:10.1126/science.abl4896\n",
    "\n",
    "[4] Uhlén M et al. \"Tissue-based map of the human proteome.\" Science (New York, N.Y.) vol. 347,6220 (2015): 1260419. https://doi.org/10.1126/science.1260419\n",
    "\n",
    "[5] Kim, Min-Sik et al. “A draft map of the human proteome.” Nature vol. 509,7502 (2014): 575-81. https://doi.org/10.1038/nature13302\n",
    "\n",
    "[6] Jiang, Lihua et al. “A Quantitative Proteome Map of the Human Body.” Cell vol. 183,1 (2020): 269-283.e19. https://doi.org/10.1016/j.cell.2020.08.036\n",
    "\n",
    "[7] Tabula Muris Consortium et al. “Single-cell transcriptomics of 20 mouse organs creates a Tabula Muris.” Nature vol. 562,7727 (2018): 367-372. doi:10.1038/s41586-018-0590-4\n",
    "\n",
    "[8] Ghandi, Mahmoud et al. “Next-generation characterization of the Cancer Cell Line Encyclopedia.” Nature vol. 569,7757 (2019): 503-508. doi:10.1038/s41586-019-1186-3\n",
    "        \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "vscode": {
   "interpreter": {
    "hash": "640e7826e3db6618de4c1a35509d73c188e8c0fc52fc519d75184b2277dca4fd"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
