{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61dc43ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90ce0804",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Tumor Gene Target Screener',\n",
    ") %}\n",
    "\n",
    "{% do DescriptionField(\n",
    "    name='data_file_description',\n",
    "    text='''\n",
    "    Files should be a tsv/csv of the form:<br />\n",
    "    <table class=\"table\">\n",
    "    <tr>\n",
    "      <td>&nbsp;</td>\n",
    "      <th>Replicate 1</th>\n",
    "      <th>Replicate 2</th>\n",
    "      <th>...</th>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>Gene|Transcript 1</th>\n",
    "      <td>0</td>\n",
    "      <td>200</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>Gene|Transcript 2</th>\n",
    "      <td>5</td>\n",
    "      <td>180</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>...</th>\n",
    "      <td>...</td>\n",
    "      <td>...</td>\n",
    "      <td>...</td>\n",
    "    </tr>\n",
    "    </table>''',\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set file = FileField(\n",
    "    name='patient_expression',\n",
    "    label='Patient RNA-seq expression vectors',\n",
    "    description='Gene/Transcripts on the rows, replicates on the columns',\n",
    "    default='GSE49155-lung-squamous-cell-carcinoma.tsv',\n",
    "    required=True,\n",
    "    examples={\n",
    "        'GSE49155-lung-squamous-cell-carcinoma.tsv': 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/GSE49155-patient.tsv',\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set background = TabField(\n",
    "    name='background',\n",
    "    label='Normal tissue background',\n",
    "    description='Patient expression will be contrasted against this background',\n",
    "    default='Precomputed',\n",
    "    choices={\n",
    "        'Precomputed': [\n",
    "            ChoiceField(\n",
    "                name='background_dataset',\n",
    "                label='Normal tissue background',\n",
    "                description='Choose Gene or Transcript Background variant if your patient vectors are at the Gene or Transcript level.',\n",
    "                choices={\n",
    "                    'GTEx - Gene': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/gtex-gene-stats.tsv\"',\n",
    "                    'GTEx - Transcript': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/gtex-transcript-stats.tsv\"',\n",
    "                    'ARCHS4 Anatomy - Gene': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/archs4-gene-anatomy-stats.tsv\"',\n",
    "                    'ARCHS4 Anatomy - Transcript': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/archs4-transcript-anatomy-stats.tsv\"',\n",
    "                    'ARCHS4 Extra - Gene': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/archs4-gene-extra-stats.tsv\"',\n",
    "                    'ARCHS4 Extra - Transcript': '\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/archs4-transcript-extra-stats.tsv\"',\n",
    "                },\n",
    "                default='GTEx - Gene',\n",
    "            ),\n",
    "        ],\n",
    "        'Custom': [\n",
    "            FileField(\n",
    "                name='background_upload',\n",
    "                label='Normal tissue background',\n",
    "                description='Given a matrix (d), genes or transcripts by samples, this matrix can be constructed with `d.T.groupby(sample_tissue_mappings).description().T`',\n",
    "                examples={\n",
    "                    'GTEx': 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/gtex-gene-stats.tsv',\n",
    "                },\n",
    "                default=None,\n",
    "                section='primary',\n",
    "            ),\n",
    "            BoolField(\n",
    "                name='transcript_level',\n",
    "                label='Whether this file is at the level of transcripts or genes',\n",
    "                default=False,\n",
    "                yes_label='Transcript Level',\n",
    "                no_label='Gene Level',\n",
    "            ),\n",
    "            DescriptionField(\n",
    "                name='background_file_description',\n",
    "                text='''\n",
    "                Files should be a tsv/csv of the form:<br />\n",
    "                <table class=\"table\">\n",
    "                <tr>\n",
    "                  <td>&nbsp;</td>\n",
    "                  <td>&nbsp;</td>\n",
    "                  <th>Tissue 1</th>\n",
    "                  <th>...</th>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>Gene|Transcript 1</th>\n",
    "                  <th>25%</th>\n",
    "                  <td>0</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>...</th>\n",
    "                  <th>...</th>\n",
    "                  <td>...</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                <tr>\n",
    "                  <th>Gene|Transcript n</th>\n",
    "                  <th>mean</th>\n",
    "                  <td>180</td>\n",
    "                  <td>...</td>\n",
    "                </tr>\n",
    "                </table>''',\n",
    "            ),\n",
    "        ],\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set background_dataset = background.value[0] %}\n",
    "{% if background.raw_value == 'Precomputed' %}\n",
    "{% set transcript_level = background.value[0].raw_value.endswith('- Transcript') %}\n",
    "{% else %}\n",
    "{% set transcript_level = background.value[1].raw_value or False %}\n",
    "{% endif %}\n",
    "\n",
    "{% set membrane_screener = BoolField(\n",
    "    name='membrane_screener',\n",
    "    label='Prioritize membrane genes',\n",
    "    description='Use membranome to identify membrane',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set normalize_to_background = BoolField(\n",
    "    name='normalize_to_background',\n",
    "    label='Normalize to background distribution',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set proteomics_vis = BoolField(\n",
    "    name='proteomics_vis',\n",
    "    label='Show protein expression profiles of gene candidates',\n",
    "    description='View protein expression levels in normal tissues, from Human Proteome Map and Human Protein Atlas proteomics data',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a8aebe2",
   "metadata": {},
   "source": [
    "# Overexpressed Candidate Identification\n",
    "\n",
    "This appyter uses RNA-seq expression data for a patient and identifies over-expressed proteins verses a baseline dataset of normal tissues such those in GTEx or ARCHS4. It then prioritizes candidates by significance and targetability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db3e09e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "import qnorm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.graph_objects as go\n",
    "from IPython.display import HTML, display, Markdown\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from maayanlab_bioinformatics.dge import limma_voom_differential_expression\n",
    "{% if not transcript_level %}\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "lookup = ncbi_genes_lookup()\n",
    "{% endif %}\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest='/dev/null'):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecd421bd",
   "metadata": {},
   "source": [
    "## Load Patient RNA-seq Expression Data\n",
    "\n",
    "Load RNA-seq expression data for the patient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9089e2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_table(filename):\n",
    "    if filename.endswith('.tsv') or filename.endswith('.tsv.gz'):\n",
    "        return pd.read_csv(filename, sep='\\t', index_col=0)\n",
    "    elif filename.endswith('.csv') or filename.endswith('.csv.gz'):\n",
    "        return pd.read_csv(filename, sep=',', index_col=0)\n",
    "    elif filename.endswith('.gct') or filename.endswith('.gct.gz'):\n",
    "        return pd.read_csv(filename, sep='\\t', index_col=0, skiprows=2)\n",
    "    else:\n",
    "        return pd.read_table(filename, sep=None, engine='python', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2772fdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "df_expr = read_table({{ file }})\n",
    "{% if transcript_level %}\n",
    "df_expr_transcripts = df_expr.index.map(lambda idx: idx.partition('.')[0])\n",
    "df_expr = df_expr.groupby(df_expr_transcripts, observed=True).sum()\n",
    "{% else %}\n",
    "df_expr_genes = df_expr.index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "df_expr = df_expr.groupby(df_expr_genes, observed=True).median()\n",
    "{% endif %}\n",
    "df_expr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d5ae5b",
   "metadata": {},
   "source": [
    "## Load Background Dataset\n",
    "\n",
    "The background dataset contains expression for normal tissues in many patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26dc8bd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "df_bg_stats = pd.read_csv({{ background_dataset }}, sep='\\t', index_col=[0,1])\n",
    "{% if transcript_level %}\n",
    "df_bg_transcripts = df_bg_stats.unstack().index.map(lambda idx: idx.partition('.')[0])\n",
    "df_bg_stats = df_bg_stats.unstack().groupby(df_bg_transcripts, observed=True).sum().stack()\n",
    "{% else %}\n",
    "df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "{% endif %}\n",
    "df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "df_bg_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "208bacc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if membrane_screener.raw_value and not transcript_level %}\n",
    "\n",
    "## Load Membrane Proteins for Screening\n",
    "\n",
    "Membrane proteins are ideal targets, we can get these from [Membranome](https://membranome.org/) among other places.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "860c071b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if membrane_screener.raw_value and not transcript_level %}\n",
    "proteins = pd.read_csv('https://lomize-group-membranome.herokuapp.com/proteins?fileFormat=csv')\n",
    "proteins = proteins[proteins['species_name_cache'] == 'Homo sapiens']\n",
    "membrane_proteins = proteins['genename'].map(lookup).dropna()\n",
    "membrane_proteins\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "893dd26a",
   "metadata": {},
   "source": [
    "## Distribution matching between patient sample(s) & the background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93796f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax11, ax12), (ax21, ax22)) = plt.subplots(2, 2)\n",
    "log2_normalize(df_expr).median(axis=1).hist(bins=100, ax=ax11)\n",
    "ax11.set_title('Median Expression')\n",
    "ax11.set_ylabel('Patient')\n",
    "df_expr.median(axis=0).hist(bins=100, ax=ax12)\n",
    "ax12.set_title('Median Sample Expression')\n",
    "log2_normalize(df_bg_expr).median(axis=1).hist(bins=100, ax=ax21)\n",
    "ax21.set_ylabel('Background')\n",
    "df_bg_expr.median(axis=0).hist(bins=100, ax=ax22)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ce3cea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_index = list(set(df_expr.index) & set(df_bg_expr.index))\n",
    "venn2([set(df_expr.index), set(df_bg_expr.index)],\n",
    "      ['Patients  ', '  Background'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6acfe2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if normalize_to_background.raw_value %}\n",
    "target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "df_expr_norm = qnorm.quantile_normalize(df_expr.loc[common_index, :], target=target_distribution)\n",
    "df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "\n",
    "fig, ((ax11, ax12), (ax21, ax22)) = plt.subplots(2, 2)\n",
    "log2_normalize(df_expr_norm).median(axis=1).hist(bins=100, ax=ax11)\n",
    "ax11.set_title('Median Expression')\n",
    "ax11.set_ylabel('Patient')\n",
    "df_expr_norm.median(axis=0).hist(bins=100, ax=ax12)\n",
    "ax12.set_title('Median Sample Expression')\n",
    "log2_normalize(df_bg_expr_norm).median(axis=1).hist(bins=100, ax=ax21)\n",
    "ax21.set_ylabel('Background')\n",
    "df_bg_expr_norm.median(axis=0).hist(bins=100, ax=ax22)\n",
    "{% else %}\n",
    "print('Warning: Proceeding without normalization')\n",
    "df_expr_norm = df_expr.loc[common_index, :]\n",
    "df_bg_expr_norm = df_bg_expr.loc[common_index, :]\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f676710a",
   "metadata": {},
   "source": [
    "## Perform Differential Expression between Patient & Background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2e4604",
   "metadata": {},
   "outputs": [],
   "source": [
    "with suppress_output():\n",
    "    dge = limma_voom_differential_expression(\n",
    "        df_bg_expr_norm, df_expr_norm,\n",
    "        voom_design=True,\n",
    "    )\n",
    "dge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26539af2",
   "metadata": {},
   "source": [
    "## Narrow Down Candidate Set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0435eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "dge['-log(adj.P.Val)'] = -np.log(dge['adj.P.Val'])\n",
    "prod = (np.abs(dge['t']) * dge['logFC'])\n",
    "dge['is_significant'] = prod > prod.mean() + 3 * prod.std()\n",
    "dge['score'] = dge['is_significant'].astype(int)\n",
    "#\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    name='Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='black',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge[dge.is_significant].index,\n",
    "    name='Significant',\n",
    "    marker=dict(\n",
    "        color='red',\n",
    "    )\n",
    "))\n",
    "fig.update_layout(\n",
    "    title='Background vs Patient Differential Expression',\n",
    "    xaxis_title='Log Fold Change',\n",
    "    yaxis_title='-Log[Adjusted P-Value]',\n",
    "    autosize=True,\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e38ab54b",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if membrane_screener.raw_value and not transcript_level %}\n",
    "dge['is_membrane'] = np.in1d(dge.index, membrane_proteins)\n",
    "dge['score'] = dge['score'] + dge['is_membrane'].astype(int)\n",
    "#\n",
    "fig = go.Figure()\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_membrane&~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[~dge.is_membrane&~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    name='Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='black',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[~dge.is_membrane&dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[~dge.is_membrane&dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge[~dge.is_membrane&dge.is_significant].index,\n",
    "    name='Significant Other',\n",
    "    showlegend=False,\n",
    "    marker=dict(\n",
    "        color='red',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_membrane&~dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_membrane&~dge.is_significant, '-log(adj.P.Val)'],\n",
    "    name='Membrane Protein',\n",
    "    marker=dict(\n",
    "        color='grey',\n",
    "    )\n",
    "))\n",
    "fig.add_trace(go.Scattergl(\n",
    "    mode='markers',\n",
    "    x=dge.loc[dge.is_membrane&dge.is_significant, 'logFC'],\n",
    "    y=dge.loc[dge.is_membrane&dge.is_significant, '-log(adj.P.Val)'],\n",
    "    text=dge[dge.is_membrane&dge.is_significant].index,\n",
    "    name='Significant Membrane Protein',\n",
    "    marker=dict(\n",
    "        color='orange',\n",
    "    )\n",
    "))\n",
    "fig.update_layout(\n",
    "    title='Background vs Tumor',\n",
    "    xaxis_title='Log Fold Change',\n",
    "    yaxis_title='-Log[Adjusted P-Value]',\n",
    "    autosize=True,\n",
    ")\n",
    "fig.show()\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c590a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "dge_final = dge[dge.score >= 1].sort_values(['score', 't'], ascending=False).iloc[:16]\n",
    "{% if not transcript_level %}\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "dge_final['Link'] = dge_final.index.map(lambda g: f\"<a href=\\\"https://cfde-gene-pages.cloud/gene/{g}\\\">Link</a>\")\n",
    "{% endif %}\n",
    "HTML(dge_final.to_html(notebook=True, escape=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3db14831",
   "metadata": {},
   "source": [
    "## Review Expression Levels of Selected Candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e0f9ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if proteomics_vis.raw_value %}\n",
    "Proteomics data were obtained from the [Human Protein Atlas](https://www.proteinatlas.org/about/download) (HPA) with IHC-based expression profiling, the [Human Proteome Map](https://www.humanproteomemap.org/download.php) (HPM) with MS-based expression quantification, and a [GTEx proteome project](https://doi.org/10.1016/j.cell.2020.08.036) using TMT MS. \n",
    "These datasets contain protein expression levels detected in normal tissues and cell types. Not all differentially-expressed gene candidates may be present in the data from each project (see table for which proteomics data are present/absent). Plots show expression levels (HPA), average spectral counts (HPM), or a log-transformed relative abundance (GTEx) by tissue/cell-type for each gene candidate (excluding expression levels from the HPA where the [reliability score](https://www.proteinatlas.org/about/assays+annotation) was uncertain). \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "29a11e48",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Cell magic `%%appyter` not found.\n"
     ]
    }
   ],
   "source": [
    "%%appyter code_eval\n",
    "{% if proteomics_vis.raw_value %}\n",
    "\n",
    "hpm = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/hpm.tsv\", sep='\\t', header=0, index_col=0)\n",
    "hpa = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/hpa.tsv\", sep='\\t', header=0, index_col=1)\n",
    "gtexp = pd.read_csv(\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/gtex_proteomics.tsv\", sep='\\t', header=0, index_col=0)\n",
    "gtexp['Name'] = gtexp.index.map(lambda idx: lookup(idx))\n",
    "gtexp.set_index('Name', inplace=True, drop=True)\n",
    "\n",
    "# Show available genes in each dataset\n",
    "available = pd.DataFrame({'Gene': dge_final.index, \n",
    "                          'in HPM': dge_final.index.isin(hpm.index), \n",
    "                          'in HPA': dge_final.index.isin(hpa.index),\n",
    "                          'in GTEx Proteomics': dge_final.index.isin(gtexp.index)})\n",
    "display(HTML(available.to_html(notebook=True, escape=False)))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c9c077c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "{% if proteomics_vis.raw_value %}\n",
    "import plotly.express as px\n",
    "hpa.Tissue = hpa[\"Tissue\"] + \", \" + hpa[\"Cell.type\"]\n",
    "hpa = hpa[hpa['Reliability'] != \"Uncertain\"] \n",
    "gtexp['tissue_specificity'] = gtexp.tissue_specificity.fillna('NA')\n",
    "{% endif %}\n",
    "\n",
    "for c in dge_final.index:\n",
    "    display(Markdown(f\"### {c}\"))\n",
    "    IQR = df_bg_stats.loc[(c, '75%')]-df_bg_stats.loc[(c, '25%')]\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(go.Box(\n",
    "        lowerfence=np.maximum(\n",
    "            df_bg_stats.loc[(c, 'min')],\n",
    "            df_bg_stats.loc[(c, '25%')] - (1.5*IQR),\n",
    "        ),\n",
    "        q1=df_bg_stats.loc[(c, '25%')],\n",
    "        median=df_bg_stats.loc[(c, '50%')],\n",
    "        q3=df_bg_stats.loc[(c, '75%')],\n",
    "        upperfence=np.minimum(\n",
    "            df_bg_stats.loc[(c, 'max')],\n",
    "            df_bg_stats.loc[(c, '75%')] + (1.5*IQR),\n",
    "        ),\n",
    "        mean=df_bg_stats.loc[(c, 'mean')],\n",
    "        sd=df_bg_stats.loc[(c, 'std')],\n",
    "        y=df_bg_stats.columns,\n",
    "        name='Background',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.add_trace(go.Box(\n",
    "        x=df_expr_norm.loc[c],\n",
    "        name='Patient (Normalized)',\n",
    "        orientation='h'\n",
    "    ))\n",
    "    fig.update_layout(title=c+\" (RNA-seq)\", height=1200)\n",
    "    fig.show()\n",
    "    \n",
    "    {% if proteomics_vis.raw_value %}\n",
    "    if c in gtexp.index:\n",
    "        fig = px.strip(gtexp.loc[gtexp.index == c], \n",
    "                       y=\"tissue\", x=\"value\",  \n",
    "                       orientation='h',\n",
    "                       stripmode=\"overlay\",\n",
    "                       hover_data=[\"tissue_specificity\"],\n",
    "                       height=30*len(pd.unique(gtexp.loc[gtexp.index == c, \"tissue\"])))\n",
    "        fig.add_trace(go.Box(x=gtexp.loc[gtexp.index == c]['value'],\n",
    "                             y=gtexp.loc[gtexp.index == c]['tissue'],\n",
    "                             orientation='h',\n",
    "                             marker=dict(color='#636EFA'),\n",
    "                             name=\"n > 1\"))\n",
    "        fig.update_layout(title=c+\" (GTEx Proteomics)\",\n",
    "                          autosize=True,\n",
    "                          showlegend=False)\n",
    "        fig.update_xaxes(title=\"log2(relative abundance)\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "       \n",
    "    if c in hpm.index:\n",
    "        fig = px.scatter(hpm.loc[[c]], \n",
    "                         y=\"Tissue\", x=\"value\", \n",
    "                         height=20*hpm.loc[[c]].shape[0])\n",
    "        fig.update_layout(title=c+\" (HPM)\", \n",
    "                          autosize=True)\n",
    "        fig.update_xaxes(title=\"Average Spectral Counts\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "    \n",
    "    if c in hpa.index:\n",
    "        fig = px.scatter(hpa.loc[[c]], \n",
    "                         y=\"Tissue\", x=\"Level\", \n",
    "                         category_orders={\"Level\": [\"Not detected\", \"Low\", \"Medium\", \"High\"]}, \n",
    "                         hover_data=[\"Reliability\"],  \n",
    "                         hover_name=\"Tissue\",\n",
    "                         height=20*hpa.loc[[c]].shape[0])\n",
    "        fig.update_layout(title=c+\" (HPA)\", \n",
    "                          showlegend=False, \n",
    "                          autosize=True, \n",
    "                          xaxis={'tickmode':'array', \n",
    "                                 'tickvals':[0, 1, 2, 3], \n",
    "                                 'ticktext':[\"Not detected\", \"Low\", \"Medium\", \"High\"]})\n",
    "        fig.update_xaxes(title=\"Tissue Expression Level\")\n",
    "        fig.update_yaxes(title=None)\n",
    "        fig.show()\n",
    "    {% endif %}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
