{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "{% do SectionField(\n",
    "    name='PRIMARY',\n",
    "    title='Upload gene sets',\n",
    "    subtitle='Upload your up and down gene sets',\n",
    "    img='file-upload.png'\n",
    ") %}\n",
    "{% do SectionField(\n",
    "    name='ENRICHMENT',\n",
    "    title='Choose Drugmonizome and Enrichr Libraries for Enrichment',\n",
    "    subtitle='Choose libraries.',\n",
    "    img='find-replace.png'\n",
    "    \n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "{% set title = StringField(\n",
    "    name='title',\n",
    "    label='Notebook name',\n",
    "    default='LDP3 Consensus Perturbagens',\n",
    "    section=\"PRIMARY\",\n",
    ") %}\n",
    "\n",
    "# {{ title.raw_value }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import math\n",
    "import time\n",
    "import requests\n",
    "import pandas as pd\n",
    "import json\n",
    "import scipy.stats as st\n",
    "from IPython.display import display, IFrame, Markdown, HTML\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from umap import UMAP\n",
    "from sklearn.manifold import TSNE\n",
    "from maayanlab_bioinformatics.normalization import quantile_normalize, zscore_normalize\n",
    "from maayanlab_bioinformatics.harmonization import ncbi_genes_lookup\n",
    "from tqdm import tqdm\n",
    "import plotly.express as px\n",
    "import numpy as np\n",
    "from matplotlib.ticker import MaxNLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "METADATA_API = \"https://ldp3.cloud/metadata-api\"\n",
    "DATA_API = \"https://ldp3.cloud/data-api/api/v1\"\n",
    "CLUSTERGRAMMER_URL = 'https://maayanlab.cloud/clustergrammer/matrix_upload/'\n",
    "S3_PREFIX = \"https://appyters.maayanlab.cloud/storage/LDP3Consensus/\"\n",
    "drugmonizome_meta_api = \"https://maayanlab.cloud/drugmonizome/metadata-api\"\n",
    "drugmonizome_data_api = \"https://maayanlab.cloud/drugmonizome/data-api/api/v1\"\n",
    "enrichr_api = 'https://maayanlab.cloud/Enrichr/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_lookup = ncbi_genes_lookup('Mammalia/Homo_sapiens')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = 1\n",
    "figure = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "def convert_genes(up_genes=[], down_genes=[]):\n",
    "    try:\n",
    "        payload = {\n",
    "           \"filter\": {\n",
    "               \"where\": {\n",
    "                   \"meta.symbol\": {\"inq\": up_genes + down_genes}\n",
    "               }\n",
    "           }\n",
    "        }\n",
    "        timeout = 0.5\n",
    "        for i in range(5):\n",
    "            res = requests.post(METADATA_API + \"/entities/find\", json=payload)\n",
    "            if res.ok:\n",
    "                break\n",
    "            else:\n",
    "                time.sleep(timeout)\n",
    "                if res.status_code >= 500:\n",
    "                    timeout = timeout * 2\n",
    "        else:\n",
    "            raise Exception(res.text)\n",
    "        results = res.json()\n",
    "        up = set(up_genes)\n",
    "        down = set(down_genes)\n",
    "        if len(up_genes) == 0 or len(down_genes) == 0:\n",
    "            converted = {\n",
    "                \"entities\": [],\n",
    "            }\n",
    "        else:\n",
    "            converted = {\n",
    "                \"up_entities\": [],\n",
    "                \"down_entities\": []\n",
    "            }\n",
    "        for i in results:\n",
    "            symbol = i[\"meta\"][\"symbol\"]\n",
    "            if \"entities\" in converted:\n",
    "                converted[\"entities\"].append(i[\"id\"])\n",
    "            elif symbol in up:\n",
    "                converted[\"up_entities\"].append(i[\"id\"])\n",
    "            elif symbol in down:\n",
    "                converted[\"down_entities\"].append(i[\"id\"])\n",
    "        return converted\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "\n",
    "\n",
    "def signature_search(genes, library):\n",
    "    try:\n",
    "        payload = {\n",
    "            **genes,\n",
    "            \"database\": library,\n",
    "            \"limit\": 1000\n",
    "        }\n",
    "        timeout = 0.5\n",
    "        for i in range(5):\n",
    "            endpoint = \"/enrich/rank\" if \"entities\" in payload else \"/enrich/ranktwosided\"\n",
    "            res = requests.post(DATA_API + endpoint, json=payload)\n",
    "            if res.ok:\n",
    "                break\n",
    "            else:\n",
    "                time.sleep(timeout)\n",
    "                if res.status_code >= 500:\n",
    "                    timeout = timeout * 2\n",
    "        else:\n",
    "            raise Exception(res.text)\n",
    "        \n",
    "        return res.json()[\"results\"]\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "\n",
    "def resolve_rank(s, gene_set_direction):\n",
    "    try:\n",
    "        sigs = {}\n",
    "        for i in s:\n",
    "            if i[p-]\n",
    "            uid = i[\"uuid\"]\n",
    "            direction = \"up\" if i[\"zscore\"] > 0 else \"down\"\n",
    "            if direction == gene_set_direction:\n",
    "                i[\"type\"] = \"mimicker\"\n",
    "                sigs[uid] = i\n",
    "            else:\n",
    "                i[\"type\"] = \"reverser\"\n",
    "                sigs[uid] = i\n",
    "            \n",
    "        payload = {\n",
    "            \"filter\": {\n",
    "                \"where\": {\n",
    "                    \"id\": {\"inq\": list(sigs.keys())}\n",
    "                },\n",
    "                \"fields\": [\n",
    "                    \"id\",\n",
    "                    \"meta.pert_name\",\n",
    "                    \"meta.pert_type\"\n",
    "                ]\n",
    "            }\n",
    "        }\n",
    "        timeout = 0.5\n",
    "        for i in range(5):\n",
    "            res = requests.post(METADATA_API + \"/signatures/find\", json=payload)\n",
    "            if res.ok:\n",
    "                break\n",
    "            else:\n",
    "                time.sleep(timeout)\n",
    "                if res.status_code >= 500:\n",
    "                    timeout = timeout * 2\n",
    "        else:\n",
    "            raise Exception(res.text)\n",
    "        results = res.json()\n",
    "        perturbagens = {\n",
    "            \"mimickers\": {},\n",
    "            \"reversers\": {}\n",
    "        }\n",
    "        for sig in results:\n",
    "            uid = sig[\"id\"]\n",
    "            scores = sigs[uid]\n",
    "            sig[\"scores\"] = scores\n",
    "            if \"pert_name\" in sig[\"meta\"]:\n",
    "                pert_name = sig[\"meta\"][\"pert_name\"]\n",
    "                if scores[\"type\"] == \"mimicker\":\n",
    "                    if pert_name not in perturbagens[\"mimickers\"]:\n",
    "                        perturbagens[\"mimickers\"][pert_name] = 0\n",
    "                    perturbagens[\"mimickers\"][pert_name] += 1\n",
    "                elif scores[\"type\"] == \"reverser\":\n",
    "                    if pert_name not in perturbagens[\"reversers\"]:\n",
    "                        perturbagens[\"reversers\"][pert_name] = 0\n",
    "                    perturbagens[\"reversers\"][pert_name] += 1\n",
    "        return perturbagens\n",
    "\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "\n",
    "\n",
    "def resolve_ranktwosided(s):\n",
    "    try:\n",
    "        sigs = {}\n",
    "        for i in s:\n",
    "            uid = i[\"uuid\"]\n",
    "            i['z-down'] = -i['z-down']\n",
    "            i['direction-down'] = -i['direction-down']\n",
    "            if i['z-up'] > 0 and i['z-down'] > 0:\n",
    "                i[\"type\"] = \"mimicker\"\n",
    "                sigs[uid] = i\n",
    "            elif i['z-up'] < 0 and i['z-down'] < 0:\n",
    "                i[\"type\"] = \"reverser\"\n",
    "                sigs[uid] = i\n",
    "            \n",
    "        payload = {\n",
    "            \"filter\": {\n",
    "                \"where\": {\n",
    "                    \"id\": {\"inq\": list(sigs.keys())}\n",
    "                },\n",
    "                \"fields\": [\n",
    "                    \"id\",\n",
    "                    \"meta.pert_name\",\n",
    "                    \"meta.pert_type\"\n",
    "                ]\n",
    "            }\n",
    "        }\n",
    "        timeout = 0.5\n",
    "        for i in range(5):\n",
    "            res = requests.post(METADATA_API + \"/signatures/find\", json=payload)\n",
    "            if res.ok:\n",
    "                break\n",
    "            else:\n",
    "                time.sleep(timeout)\n",
    "                if res.status_code >= 500:\n",
    "                    timeout = timeout * 2\n",
    "        else:\n",
    "            raise Exception(res.text)\n",
    "        results = res.json()\n",
    "        perturbagens = {\n",
    "            \"mimickers\": {},\n",
    "            \"reversers\": {}\n",
    "        }\n",
    "        for sig in results:\n",
    "            uid = sig[\"id\"]\n",
    "            scores = sigs[uid]\n",
    "            sig[\"scores\"] = scores\n",
    "            if \"pert_name\" in sig[\"meta\"]:\n",
    "                pert_name = sig[\"meta\"][\"pert_name\"]\n",
    "                if scores[\"type\"] == \"mimicker\":\n",
    "                    if pert_name not in perturbagens[\"mimickers\"]:\n",
    "                        perturbagens[\"mimickers\"][pert_name] = 0\n",
    "                    perturbagens[\"mimickers\"][pert_name] += 1\n",
    "                elif scores[\"type\"] == \"reverser\":\n",
    "                    if pert_name not in perturbagens[\"reversers\"]:\n",
    "                        perturbagens[\"reversers\"][pert_name] = 0\n",
    "                    perturbagens[\"reversers\"][pert_name] += 1\n",
    "        return perturbagens\n",
    "\n",
    "    except Exception as e:\n",
    "        print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clustergrammer(df, name, figure, label=\"Clustergrammer\"):\n",
    "    clustergram_df = df.rename(columns={i:\"Signature: %s\"%i for i in df.columns}, index={i:\"Drug: %s\"%i for i in df.index})\n",
    "    clustergram_df.to_csv(name, sep=\"\\t\")\n",
    "    response = ''\n",
    "    timeout = 0.5\n",
    "    for i in range(5):\n",
    "        try:\n",
    "            res = requests.post(CLUSTERGRAMMER_URL, files={'file': open(name, 'rb')})\n",
    "            if not res.ok:\n",
    "                response = res.text\n",
    "                time.sleep(timeout)\n",
    "                if res.status_code >= 500:\n",
    "                    timeout = timeout * 2\n",
    "            else:\n",
    "                clustergrammer_url = res.text.replace(\"http:\",\"https:\")   \n",
    "                break\n",
    "        except Exception as e:\n",
    "            response = e\n",
    "            time.sleep(2)\n",
    "    else:\n",
    "        if type(response) == Exception:\n",
    "            raise response\n",
    "        else:\n",
    "            raise Exception(response)\n",
    "    print(clustergrammer_url)\n",
    "    display(IFrame(clustergrammer_url, width=\"1000\", height=\"1000\"))\n",
    "    display(Markdown(\"**Figure %d** %s [Go to url](%s)\"%(figure, label, clustergrammer_url)))\n",
    "    figure += 1\n",
    "    return figure\n",
    "\n",
    "cmap = sns.cubehelix_palette(50, hue=0.05, rot=0, light=1, dark=0)\n",
    "\n",
    "def heatmap(df, filename, figure, label, width=15, height=15):\n",
    "    fig = plt.figure(figsize=(width,height))\n",
    "    cg = sns.clustermap(df, cmap=cmap, figsize=(width, height))\n",
    "    cg.ax_row_dendrogram.set_visible(False)\n",
    "    cg.ax_col_dendrogram.set_visible(False)\n",
    "    display(cg)\n",
    "    plt.show()\n",
    "    cg.savefig(filename)\n",
    "    display(Markdown(\"**Figure %d** %s\"%(figure, label)))\n",
    "    figure+=1\n",
    "    return figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "{% set up_gene_sets = FileField(\n",
    "    name='up_gene_sets',\n",
    "    label='up gene sets',\n",
    "    default='covid19_up.gmt',\n",
    "    section=\"PRIMARY\",\n",
    "    examples={\n",
    "        'covid19_up.gmt': 'https://appyters.maayanlab.cloud/storage/LDP3Consensus/covid19_up.gmt'\n",
    "    }\n",
    ") %}\n",
    "\n",
    "{% set down_gene_sets = FileField(\n",
    "    name='down_gene_sets',\n",
    "    label='down gene sets',\n",
    "    default='covid19_down.gmt',\n",
    "    section=\"PRIMARY\",\n",
    "    examples={\n",
    "        'covid19_down.gmt': 'https://appyters.maayanlab.cloud/storage/LDP3Consensus/covid19_down.gmt'\n",
    "    }\n",
    ") %}\n",
    "\n",
    "{% set input_meta = FileField(\n",
    "    name='input_meta',\n",
    "    label='Metadata File',\n",
    "    default='covid19_meta.tsv',\n",
    "    section=\"PRIMARY\",\n",
    "    examples={\n",
    "        'covid19_meta.tsv': 'https://appyters.maayanlab.cloud/storage/LDP3Consensus/covid19_meta.tsv',\n",
    "    }\n",
    ") %}\n",
    "{% set color_by =  StringField(name='group_by', label='Group By', description=\"Group By Metadata\", default='', section='PRIMARY')%}\n",
    "\n",
    "\n",
    "up_gene_sets = {{ up_gene_sets }}\n",
    "down_gene_sets = {{ down_gene_sets }}\n",
    "input_meta = {{ input_meta }}\n",
    "color_by = {{ color_by }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set_direction = None\n",
    "if up_gene_sets == '':\n",
    "    gene_set_direction = \"down\"\n",
    "    print(\"Up gene sets was not uploaded. Gene set direction is set to down.\")\n",
    "elif down_gene_sets == '':\n",
    "    gene_set_direction = \"up\"\n",
    "    print(\"Down gene sets was not uploaded. Gene set direction is set to up.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df = pd.read_csv(input_meta, sep=\"\\t\", index_col=0)\n",
    "display(meta_df.head())\n",
    "display(Markdown(\"**Table %d** Input Metadata\"%table))\n",
    "table+=1\n",
    "if color_by == \"\":\n",
    "    color_by = meta_df.columns[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "alpha = {{FloatField(name='alpha', label='p-value cutoff', default=0.05, section='PRIMARY')}}\n",
    "perc = {{FloatField(name='perc',\n",
    "                         label='percentage',\n",
    "                         description='Minimum percentage cutoff for perturbagen instances in signatures',\n",
    "                         default=0.1, section='PRIMARY')}}\n",
    "top_perts = {{IntField(name='top_perts', label='top perturbation', default=50, section='PRIMARY')}}\n",
    "consensus_method = {{ ChoiceField(\n",
    "  name='consensus_method',\n",
    "  label='Consensus method',\n",
    "  description='Please select a method for getting the consensus',\n",
    "  default='zscore',\n",
    "  choices={\n",
    "    'zscore': \"'zscore'\",\n",
    "    'top count': \"'count'\",\n",
    "  },\n",
    "  section='PRIMARY') }}\n",
    "perplexity = {{IntField(name='perplexity', label='Perplexity', description=\"t-SNE perplexity\", default=15, section='PRIMARY')}}\n",
    "n_neighbors = {{IntField(name='n_neighbors', label='n_neighbors', description=\"UMAP's n_neighbors\", default=15, section='PRIMARY')}}\n",
    "random_state = {{IntField(name='random_state', label='Random State', description=\"Random State\", default=21, section='PRIMARY')}}\n",
    "\n",
    "width = {{FloatField(name='width', label='image width', default=10, section='PRIMARY')}}\n",
    "height = {{FloatField(name='height', label='image height', default=10, section='PRIMARY')}}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "signatures = {}\n",
    "if not up_gene_sets == '':\n",
    "    with open(up_gene_sets) as upfile:\n",
    "        for line in upfile:\n",
    "            unpacked = line.strip().split(\"\\t\\t\")\n",
    "            if not len(unpacked) == 2:\n",
    "                raise ValueError(\"GMT is not formatted properly, please consult the README of the appyter for proper formatting\")\n",
    "            sigid, geneset_str = unpacked\n",
    "            genes = []\n",
    "            for i in geneset_str.split(\"\\t\"):\n",
    "                gene = i.split(\",\")[0]\n",
    "                gene_name = ncbi_lookup(gene.upper())\n",
    "                if gene_name:\n",
    "                    genes.append(gene_name)\n",
    "            signatures[sigid] = {\n",
    "                \"up_genes\": genes,\n",
    "                \"down_genes\": []\n",
    "            }\n",
    "if not down_gene_sets == '':\n",
    "    with open(down_gene_sets) as downfile:\n",
    "        for line in downfile:\n",
    "            unpacked = line.strip().split(\"\\t\\t\")\n",
    "            if not len(unpacked) == 2:\n",
    "                raise ValueError(\"GMT is not formatted properly, please consult the README of the appyter for proper formatting\")\n",
    "            sigid, geneset_str = unpacked\n",
    "            if sigid not in signatures and gene_set_direction == None:\n",
    "                raise ValueError(\"%s did not match any of the up signatures, make sure that the signature names are the same for both up and down genes\"%sigid)\n",
    "            else:\n",
    "                genes = []\n",
    "                for i in geneset_str.split(\"\\t\"):\n",
    "                    gene = i.split(\",\")[0]\n",
    "                    gene_name = ncbi_lookup(gene)\n",
    "                    if gene_name:\n",
    "                        genes.append(gene_name)\n",
    "                if sigid in signatures:\n",
    "                    signatures[sigid][\"down_genes\"] = genes\n",
    "                else:\n",
    "                    signatures[sigid] = {\n",
    "                        \"up_genes\": [],\n",
    "                        \"down_genes\": genes\n",
    "                    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "datasets = {{ MultiChoiceField(name='datasets',\n",
    "                                description='Select the LINCS Dataset that you want to use for Signature Search',\n",
    "                                label='LINCS Datasets',\n",
    "                                default=[\n",
    "                                    \"LINCS L1000 CRISPR Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 Chemical Perturbations (2021)\",\n",
    "                                ],\n",
    "                                section = 'PRIMARY',\n",
    "                                choices=[\n",
    "                                    \"LINCS L1000 Antibody Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 Ligand Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 Overexpression Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 CRISPR Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 shRNA Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 Chemical Perturbations (2021)\",\n",
    "                                    \"LINCS L1000 siRNA Perturbations (2021)\",\n",
    "                                ]\n",
    "                              )                           \n",
    "}}\n",
    "\n",
    "drugmonizome_datasets = {{ MultiChoiceField(name='drugmonizome_datasets',\n",
    "                                description='Select the drugmonizome libraries to perform enrichment analysis for consensus drugs',\n",
    "                                label='Drugmonizome Libraries',\n",
    "                                default=[\"L1000FWD_GO_Biological_Processes_drugsetlibrary_up\", \"L1000FWD_GO_Biological_Processes_drugsetlibrary_down\"],\n",
    "                                section = 'ENRICHMENT',\n",
    "                                choices=[\n",
    "                                    \"KinomeScan_kinase_drugsetlibrary\",\n",
    "                                    \"L1000FWD_GO_Cellular_Component_drugsetlibrary_up\",\n",
    "                                    \"L1000FWD_KEGG_Pathways_drugsetlibrary_down\",\n",
    "                                    \"L1000FWD_signature_drugsetlibrary_up\",\n",
    "                                    \"Geneshot_associated_drugsetlibrary\",\n",
    "                                    \"Geneshot_predicted_generif_drugsetlibrary\",\n",
    "                                    \"SIDER_indications_drugsetlibrary\",\n",
    "                                    \"L1000FWD_GO_Molecular_Function_drugsetlibrary_up\",\n",
    "                                    \"L1000FWD_GO_Molecular_Function_drugsetlibrary_down\",\n",
    "                                    \"L1000FWD_KEGG_Pathways_drugsetlibrary_up\",\n",
    "                                    \"SIDER_side_effects_drugsetlibrary\",\n",
    "                                    \"DrugRepurposingHub_target_drugsetlibrary\",\n",
    "                                    \"L1000FWD_GO_Biological_Processes_drugsetlibrary_down\",\n",
    "                                    \"L1000FWD_GO_Biological_Processes_drugsetlibrary_up\",\n",
    "                                    \"L1000FWD_GO_Cellular_Component_drugsetlibrary_down\",\n",
    "                                    \"ATC_drugsetlibrary\",\n",
    "                                    \"Drugbank_smallmolecule_target_drugsetlibrary\",\n",
    "                                    \"STITCH_target_drugsetlibrary\",\n",
    "                                    \"Geneshot_predicted_autorif_drugsetlibrary\",\n",
    "                                    \"Drugbank_smallmolecule_enzyme_drugsetlibrary\",\n",
    "                                    \"PharmGKB_OFFSIDES_side_effects_drugsetlibrary\",\n",
    "                                    \"CREEDS_signature_drugsetlibrary_down\",\n",
    "                                    \"Geneshot_predicted_tagger_drugsetlibrary\",\n",
    "                                    \"RDKIT_maccs_fingerprints_drugsetlibrary\",\n",
    "                                    \"CREEDS_signature_drugsetlibrary_up\",\n",
    "                                    \"DrugCentral_target_drugsetlibrary\",\n",
    "                                    \"L1000FWD_signature_drugsetlibrary_down\",\n",
    "                                    \"L1000FWD_predicted_side_effects\",\n",
    "                                    \"Drugbank_smallmolecule_carrier_drugsetlibrary\",\n",
    "                                    \"PubChem_fingerprints_drugsetlibrary\",\n",
    "                                    \"Geneshot_predicted_enrichr_drugsetlibrary\",\n",
    "                                    \"DrugRepurposingHub_moa_drugsetlibrary\",\n",
    "                                    \"Geneshot_predicted_coexpression_drugsetlibrary\",\n",
    "                                    \"Drugbank_smallmolecule_transporter_drugsetlibrary\",\n",
    "                                    \"PharmGKB_snp_drugsetlibrary\"\n",
    "                                ]\n",
    "                              )                           \n",
    "}}\n",
    "\n",
    "transcription_libraries = {{ MultiChoiceField(name='transcription_libraries', \n",
    "                                            description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                              label='Transcription', \n",
    "                                              default=[], \n",
    "                                              section = 'ENRICHMENT',\n",
    "                                              choices=[\n",
    "                                                'ARCHS4_TFs_Coexp',\n",
    "                                                'ChEA_2016',\n",
    "                                                'ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X',\n",
    "                                                'ENCODE_Histone_Modifications_2015',\n",
    "                                                'ENCODE_TF_ChIP-seq_2015',\n",
    "                                                'Epigenomics_Roadmap_HM_ChIP-seq',\n",
    "                                                'Enrichr_Submissions_TF-Gene_Coocurrence',\n",
    "                                                'Genome_Browser_PWMs',\n",
    "                                                'lncHUB_lncRNA_Co-Expression',\n",
    "                                                'miRTarBase_2017',\n",
    "                                                'TargetScan_microRNA_2017',\n",
    "                                                'TF-LOF_Expression_from_GEO',\n",
    "                                                'TF_Perturbations_Followed_by_Expression',\n",
    "                                                'Transcription_Factor_PPIs',\n",
    "                                                'TRANSFAC_and_JASPAR_PWMs',\n",
    "                                                'TRRUST_Transcription_Factors_2019']) \n",
    "                           }}\n",
    "\n",
    "\n",
    "pathways_libraries = {{ MultiChoiceField(name='pathways_libraries',\n",
    "                                         description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                         label='Pathways',\n",
    "                                         default=[],\n",
    "                                         section = 'ENRICHMENT',\n",
    "                                         choices=[\n",
    "                                            'ARCHS4_Kinases_Coexp',\n",
    "                                            'BioCarta_2016',\n",
    "                                            'BioPlanet_2019',\n",
    "                                            'BioPlex_2017',\n",
    "                                            'CORUM',\n",
    "                                            'Elsevier_Pathway_Collection',\n",
    "                                            'HMS_LINCS_KinomeScan',\n",
    "                                            'HumanCyc_2016',\n",
    "                                            'huMAP',\n",
    "                                            'KEA_2015',\n",
    "                                            'KEGG_2019_Human',\n",
    "                                            'KEGG_2019_Mouse',\n",
    "                                            'Kinase_Perturbations_from_GEO_down',\n",
    "                                            'Kinase_Perturbations_from_GEO_up',\n",
    "                                            'L1000_Kinase_and_GPCR_Perturbations_down',\n",
    "                                            'L1000_Kinase_and_GPCR_Perturbations_up',\n",
    "                                            'NCI-Nature_2016',\n",
    "                                            'NURSA_Human_Endogenous_Complexome',\n",
    "                                            'Panther_2016',\n",
    "                                            'Phosphatase_Substrates_from_DEPOD',\n",
    "                                            'PPI_Hub_Proteins',\n",
    "                                            'Reactome_2016',\n",
    "                                            'SILAC_Phosphoproteomics',\n",
    "                                            'SubCell_BarCode',\n",
    "                                            'Virus-Host_PPI_P-HIPSTer_2020',\n",
    "                                            'WikiPathways_2019_Human',\n",
    "                                            'WikiPathways_2019_Mouse']) \n",
    "                      }}    \n",
    "    \n",
    "  \n",
    "ontologies_libraries = {{ MultiChoiceField(name='ontologies_libraries', \n",
    "                                           description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                           label='Ontologies',\n",
    "                                           default=['GO_Biological_Process_2018'],\n",
    "                                           section = 'ENRICHMENT',\n",
    "                                           choices=[\n",
    "                                            'GO_Biological_Process_2018',\n",
    "                                            'GO_Cellular_Component_2018',\n",
    "                                            'GO_Molecular_Function_2018',\n",
    "                                            'Human_Phenotype_Ontology',\n",
    "                                            'Jensen_COMPARTMENTS',\n",
    "                                            'Jensen_DISEASES',\n",
    "                                            'Jensen_TISSUES',\n",
    "                                            'MGI_Mammalian_Phenotype_Level_4_2019']) \n",
    "                        }} \n",
    "\n",
    "    \n",
    "diseases_drugs_libraries = {{ MultiChoiceField(name='diseases_drugs_libraries',\n",
    "                                               description='Select the Enrichr libraries you would like in your figure.',\n",
    "                                               label='Diseases/Drugs',\n",
    "                                               default=[],\n",
    "                                               section = 'ENRICHMENT',\n",
    "                                               choices=[    \n",
    "                                                    'Achilles_fitness_decrease',\n",
    "                                                    'Achilles_fitness_increase',\n",
    "                                                    'ARCHS4_IDG_Coexp',\n",
    "                                                    'ClinVar_2019',\n",
    "                                                    'dbGaP',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Broad_CellLines_2019',\n",
    "                                                    'DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019',\n",
    "                                                    'DisGeNET',\n",
    "                                                    'DrugMatrix',\n",
    "                                                    'DSigDB',\n",
    "                                                    'GeneSigDB',\n",
    "                                                    'GWAS_Catalog_2019',\n",
    "                                                    'LINCS_L1000_Chem_Pert_down',\n",
    "                                                    'LINCS_L1000_Chem_Pert_up',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_down',\n",
    "                                                    'LINCS_L1000_Ligand_Perturbations_up',\n",
    "                                                    'MSigDB_Computational',\n",
    "                                                    'MSigDB_Oncogenic_Signatures',\n",
    "                                                    'Old_CMAP_down',\n",
    "                                                    'Old_CMAP_up',\n",
    "                                                    'OMIM_Disease',\n",
    "                                                    'OMIM_Expanded',\n",
    "                                                    'PheWeb_2019',\n",
    "                                                    'Rare_Diseases_AutoRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_AutoRIF_Gene_Lists',\n",
    "                                                    'Rare_Diseases_GeneRIF_ARCHS4_Predictions',\n",
    "                                                    'Rare_Diseases_GeneRIF_Gene_Lists',\n",
    "                                                    'UK_Biobank_GWAS_v1',\n",
    "                                                    'Virus_Perturbations_from_GEO_down',\n",
    "                                                    'Virus_Perturbations_from_GEO_up',\n",
    "                                                    'VirusMINT']) \n",
    "                            }}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichr_libraries = transcription_libraries + pathways_libraries + ontologies_libraries + diseases_drugs_libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_map = {\n",
    "  \"LINCS L1000 Antibody Perturbations (2021)\": \"l1000_aby\",\n",
    "  \"LINCS L1000 Ligand Perturbations (2021)\": \"l1000_lig\",\n",
    "  \"LINCS L1000 Overexpression Perturbations (2021)\": \"l1000_oe\",\n",
    "  \"LINCS L1000 CRISPR Perturbations (2021)\": \"l1000_xpr\",\n",
    "  \"LINCS L1000 shRNA Perturbations (2021)\": \"l1000_shRNA\",\n",
    "  \"LINCS L1000 Chemical Perturbations (2021)\": \"l1000_cp\",\n",
    "  \"LINCS L1000 siRNA Perturbations (2021)\": \"l1000_siRNA\"\n",
    "}\n",
    "\n",
    "gene_page = {\n",
    "  \"LINCS L1000 Ligand Perturbations (2021)\",\n",
    "  \"LINCS L1000 Overexpression Perturbations (2021)\",\n",
    "  \"LINCS L1000 CRISPR Perturbations (2021)\",\n",
    "  \"LINCS L1000 shRNA Perturbations (2021)\",\n",
    "  \"LINCS L1000 siRNA Perturbations (2021)\"\n",
    "}\n",
    "\n",
    "drug_page = {\n",
    "  \"LINCS L1000 Chemical Perturbations (2021)\": \"l1000_cp\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enriched = {lib:{\"mimickers\": {}, \"reversers\": {}} for lib in datasets}\n",
    "enriched = {\"mimickers\": {lib: {} for lib in datasets}, \"reversers\": {lib: {} for lib in datasets}}\n",
    "\n",
    "for k,sig in tqdm(signatures.items()):    \n",
    "    try:\n",
    "        time.sleep(0.1)\n",
    "        genes = convert_genes(sig[\"up_genes\"],sig[\"down_genes\"])\n",
    "        if (\"entities\" in genes and len(genes[\"entities\"]) > 5) or (len(genes[\"up_entities\"]) > 5 and len(genes[\"down_entities\"]) > 5):\n",
    "            for lib in datasets:\n",
    "                library = dataset_map[lib]\n",
    "                s = signature_search(genes, library)\n",
    "                if gene_set_direction == None:\n",
    "                    perturbagens = resolve_ranktwosided(s)\n",
    "                else:\n",
    "                    perturbagens = resolve_rank(s, gene_set_direction)\n",
    "                enriched[\"mimickers\"][lib][k] = perturbagens[\"mimickers\"]\n",
    "                enriched[\"reversers\"][lib][k] = perturbagens[\"reversers\"]\n",
    "                time.sleep(0.1)\n",
    "    except Exception as e:\n",
    "        print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_clickable(link):\n",
    "    # target _blank to open new window\n",
    "    # extract clickable text to display for your link\n",
    "    text = link.split('=')[1]\n",
    "    return f'<a target=\"_blank\" href=\"{link}\">{text}</a>'\n",
    "\n",
    "def metadata_plot(df, x,y):\n",
    "    colors = meta_df.columns\n",
    "    plot_rows = int(len(colors)/2) if len(colors) % 2 == 0 else int(len(colors)/2 + 1)\n",
    "    fig, axes = plt.subplots(plot_rows, 2, sharex=True, sharey=True, figsize=(20,8*plot_rows))\n",
    "    ax_1 = 0\n",
    "    ax_2 = 0\n",
    "    for color in colors:\n",
    "        if ax_2 == 2:\n",
    "            ax_2 = 0\n",
    "            ax_1+=1\n",
    "        sns.scatterplot(\n",
    "            data=df,\n",
    "            x=x, y=y,\n",
    "            hue=color,\n",
    "            ax=axes[ax_1, ax_2] if plot_rows > 1 else axes[ax_2]\n",
    "        )\n",
    "        if plot_rows > 1:\n",
    "            axes[ax_1, ax_2].set_title(color)\n",
    "        else:\n",
    "            axes[ax_2].set_title(color)\n",
    "        ax_2+=1\n",
    "    if plot_rows * 2 > len(colors):\n",
    "        if plot_rows > 1:\n",
    "            fig.delaxes(axes[ax_1, ax_2])\n",
    "        else:\n",
    "            fig.delaxes(axes[ax_2])\n",
    "    plt.show()\n",
    "\n",
    "def get_tsne(df, label, figure):\n",
    "#     perplexity = min(10, len(df.columns)-1)\n",
    "    X_embedded = TSNE(n_components=2,\n",
    "                      perplexity=perplexity,\n",
    "                      random_state=random_state,\n",
    "                     ).fit_transform(df.T)\n",
    "    tsne_df = pd.DataFrame(X_embedded, columns=[\"t-SNE 1\", \"t-SNE 2\"])\n",
    "    tsne_df['label'] = df.columns\n",
    "    if (not input_meta == \"\"):\n",
    "        tsne_df = tsne_df.merge(right=meta_df, left_on=\"label\", right_index=True)\n",
    "    if color_by:\n",
    "        display(\n",
    "            px.scatter(\n",
    "                tsne_df.loc[~tsne_df[color_by].isna()],\n",
    "                x=\"t-SNE 1\",\n",
    "                y=\"t-SNE 2\",\n",
    "                color=color_by,\n",
    "                hover_data=tsne_df.columns,\n",
    "              )\n",
    "        )\n",
    "        display(Markdown(\"**Figure %d** t-SNE plot of %s colored by %s\"%(figure, label, color_by)))\n",
    "    else:\n",
    "        px.scatter(\n",
    "            tsne_df,\n",
    "            x=\"t-SNE 1\",\n",
    "            y=\"t-SNE 2\",\n",
    "            hover_data=tsne_df.columns,\n",
    "          )\n",
    "        display(Markdown(\"**Figure %d** t-SNE plot of %s\"%(figure, label)))\n",
    "    if not input_meta == \"\" and len(meta_df.columns) > 1:\n",
    "        metadata_plot(tsne_df, 't-SNE 1', 't-SNE 2')\n",
    "        figure+=1\n",
    "        display(Markdown(\"**Figure %d** t-SNE plot of %s colored by metadata\"%(figure, label)))\n",
    "    return figure + 1\n",
    "\n",
    "def get_umap(df, label, figure):\n",
    "#     n_neighbors = min(15, len(df.columns)-1)\n",
    "    \n",
    "    consensus_umap = UMAP(\n",
    "      random_state=random_state,\n",
    "      n_neighbors=n_neighbors,\n",
    "      n_components=2,\n",
    "      metric='cosine',\n",
    "      min_dist=0.3,\n",
    "    )\n",
    "    consensus_umap.fit(df.T.values)\n",
    "    umap_df = pd.DataFrame(consensus_umap.transform(df.T.values),\n",
    "                           columns=[\"UMAP 1\", \"UMAP 2\"])\n",
    "    umap_df['label'] = df.columns\n",
    "    if (not input_meta == \"\"):\n",
    "        umap_df = umap_df.merge(right=meta_df, left_on=\"label\", right_index=True)\n",
    "#     display(app.run_server(mode='inline'))\n",
    "    if color_by:\n",
    "        display(\n",
    "            px.scatter(\n",
    "                umap_df.loc[~umap_df[color_by].isna()],\n",
    "                x=\"UMAP 1\",\n",
    "                y=\"UMAP 2\",\n",
    "                color=color_by,\n",
    "                hover_data=umap_df.columns,\n",
    "              )\n",
    "        )\n",
    "        display(Markdown(\"**Figure %d** UMAP plot of %s colored by %s\"%(figure, label, color_by)))\n",
    "    else:\n",
    "        display(\n",
    "            px.scatter(\n",
    "                umap_df,\n",
    "                x=\"UMAP 1\",\n",
    "                y=\"UMAP 2\",\n",
    "                color=color_by,\n",
    "                hover_data=umap_df.columns,\n",
    "              )\n",
    "        )\n",
    "        display(Markdown(\"**Figure %d** UMAP plot of %s\"%(figure, label)))\n",
    "    if not input_meta == \"\" and len(meta_df.columns) > 1:\n",
    "        metadata_plot(umap_df, 'UMAP 1', 'UMAP 2')\n",
    "        figure+=1\n",
    "        display(Markdown(\"**Figure %d** UMAP plot of %s colored by metadata\"%(figure, label)))\n",
    "    return figure + 1\n",
    "\n",
    "annot_dict = {}\n",
    "def bar_chart(enrichment, title=''):\n",
    "    bar_color = 'tomato'\n",
    "    bar_color_not_sig = 'lightgrey'\n",
    "    edgecolor=None\n",
    "    linewidth=0\n",
    "    if len(enrichment) > 10:\n",
    "        enrichment = enrichment[0:10]\n",
    "    enrichment_names = [i[\"name\"] for i in enrichment]\n",
    "    enrichment_scores = [i[\"pval\"] for i in enrichment]\n",
    "    plt.figure(figsize=(10,4))\n",
    "    bar_colors = [bar_color if (x < 0.05) else bar_color_not_sig for x in enrichment_scores]\n",
    "    fig = sns.barplot(x=np.log10(enrichment_scores)*-1, y=enrichment_names, palette=bar_colors, edgecolor=edgecolor, linewidth=linewidth)\n",
    "    fig.axes.get_yaxis().set_visible(False)\n",
    "    fig.set_xlabel('-Log10(p-value)',fontsize=25)\n",
    "    fig.xaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    fig.tick_params(axis='x', which='major', labelsize=20)\n",
    "    if max(np.log10(enrichment_scores)*-1)<1:\n",
    "        fig.xaxis.set_ticks(np.arange(0, max(np.log10(enrichment_scores)*-1), 0.1))\n",
    "    for ii,annot in enumerate(enrichment_names):\n",
    "        if annot in annot_dict.keys():\n",
    "            annot = annot_dict[annot]\n",
    "        if enrichment_scores[ii] < 0.05:\n",
    "            annot = '  *'.join([annot, str(str(np.format_float_scientific(enrichment_scores[ii],precision=2)))]) \n",
    "        else:\n",
    "            annot = '  '.join([annot, str(str(np.format_float_scientific(enrichment_scores[ii],precision=2)))])\n",
    "\n",
    "        title_start= max(fig.axes.get_xlim())/200\n",
    "        fig.text(title_start,ii,annot,ha='left',wrap = True, fontsize = 12)\n",
    "        fig.patch.set_edgecolor('black')  \n",
    "        fig.patch.set_linewidth('2')\n",
    "    plt.show()\n",
    "        \n",
    "\n",
    "def get_drugmonizome_plot(consensus, label, figure, dataset):\n",
    "    payload = {\n",
    "        \"filter\":{\n",
    "            \"where\": {\n",
    "                \"meta.Name\": {\n",
    "                    \"inq\": [i.lower() for i in consensus.index]\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    res = requests.post(drugmonizome_meta_api + \"/entities/find\", json=payload)\n",
    "\n",
    "    query = {\n",
    "        \"entities\": [i[\"id\"] for i in res.json()],\n",
    "        \"limit\": 1000,\n",
    "        \"database\": dataset\n",
    "    }\n",
    "\n",
    "    res = requests.post(drugmonizome_data_api + \"/enrich/overlap\", json=query)\n",
    "\n",
    "    scores = res.json()[\"results\"]\n",
    "    uids = {i[\"uuid\"]: i for i in scores}\n",
    "\n",
    "    payload = {\n",
    "        \"filter\":{\n",
    "            \"where\": {\n",
    "                \"id\": {\n",
    "                    \"inq\": list(uids.keys())\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "\n",
    "    res = requests.post(drugmonizome_meta_api + \"/signatures/find\", json=payload)\n",
    "\n",
    "    sigs = res.json()\n",
    "    sigs = res.json()\n",
    "    scores = []\n",
    "    for i in sigs:\n",
    "        score = uids[i[\"id\"]]\n",
    "        scores.append({\n",
    "            \"name\": i[\"meta\"][\"Term\"][0][\"Name\"],\n",
    "            \"pval\": score[\"p-value\"]\n",
    "        })\n",
    "    \n",
    "    scores.sort(key=lambda x: x['pval'])\n",
    "    bar_chart(scores, label)\n",
    "    display(Markdown(\"**Figure %d** %s\"%(figure, label)))\n",
    "    figure += 1\n",
    "    return figure\n",
    "\n",
    "def get_enrichr_bar(userListId, enrichr_library, figure, label):\n",
    "    query_string = '?userListId=%s&backgroundType=%s'\n",
    "    res = requests.get(\n",
    "        enrichr_api + 'enrich' + query_string % (userListId, enrichr_library)\n",
    "     )\n",
    "    if not res.ok:\n",
    "        raise Exception('Error fetching enrichment results')\n",
    "\n",
    "    data = res.json()[enrichr_library]\n",
    "    scores = [{\"name\": i[1], \"pval\": i[2]} for i in data]\n",
    "    scores.sort(key=lambda x: x['pval'])\n",
    "    bar_chart(scores, label)\n",
    "    display(Markdown(\"**Figure %d** %s\"%(figure, label)))\n",
    "    figure +=1\n",
    "    return figure\n",
    "\n",
    "def enrichment(consensus, label, figure):\n",
    "    gene_names = [i.upper() for i in consensus.index]\n",
    "    genes_str = '\\n'.join(gene_names)\n",
    "    description = label\n",
    "    payload = {\n",
    "        'list': (None, genes_str),\n",
    "        'description': (None, description)\n",
    "    }\n",
    "\n",
    "    res = requests.post(enrichr_api + 'addList', files=payload)\n",
    "    if not res.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "\n",
    "    data = res.json()\n",
    "    userListId = data[\"userListId\"]\n",
    "    for d in enrichr_libraries:\n",
    "        l = \"Enrichr %s hits for %s\"%(d, label)\n",
    "        figure = get_enrichr_bar(userListId, d, figure, l)\n",
    "    return figure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mimickers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "direction = \"mimickers\"\n",
    "for lib in datasets:\n",
    "    library = dataset_map[lib]\n",
    "    display(Markdown(\"### %s\"%lib))\n",
    "    index = set()\n",
    "    pert_dict = enriched[direction][lib]\n",
    "    for v in pert_dict.values():\n",
    "        index = index.union(v.keys())\n",
    "    df = pd.DataFrame(0, index=index, columns=pert_dict.keys())\n",
    "    for k,v in pert_dict.items():\n",
    "        for pert, cnt in v.items():\n",
    "            df.at[pert, k] += cnt\n",
    "    df = df.loc[df.sum(1).sort_values(ascending=False).index]\n",
    "    filename = \"pert_matrix_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "    df.to_csv(filename, sep=\"\\t\")\n",
    "#     display(df.head())\n",
    "#     display(Markdown(\"**Table %d** Mimicker perturbagens using %s dataset ([download](./%s))\"%\n",
    "#                      (table, lib, filename)))\n",
    "    table += 1\n",
    "    # stat_df = pd.DataFrame(0, index=df.index, columns=[\"count\", \"z-score\", \"p-value\"])\n",
    "    empirical_stat = pd.read_csv(S3_PREFIX + \"%s_%s.tsv\" % (library, direction), sep=\"\\t\", index_col=0)\n",
    "    df = df[(df>0).sum(1) > len(df.columns) * perc]\n",
    "    index = set(df.index).intersection(empirical_stat.index)\n",
    "    filtered_df = df.loc[index]    \n",
    "    if lib in gene_page:\n",
    "        stat_df = pd.DataFrame(0, index=index, columns=[\"count\", \"z-score\", \"p-value\", \"Enrichr gene page\"])\n",
    "        stat_df['count'] = filtered_df.sum(1)\n",
    "        # Compute zstat and p value\n",
    "        stat_df[\"z-score\"] = (filtered_df.mean(1) - empirical_stat[\"mean\"]) / empirical_stat[\"std\"]\n",
    "        stat_df[\"p-value\"] = stat_df['z-score'].apply(lambda x: 1-st.norm.cdf(x))\n",
    "        if consensus_method == 'zscore':\n",
    "            #Filter by p-value\n",
    "            stat_df = stat_df[stat_df[\"p-value\"]<alpha].sort_values(by=[\"z-score\"], ascending=False)\n",
    "        else:\n",
    "            stat_df = stat_df.sort_values(by=[\"count\"], ascending=False)\n",
    "\n",
    "        stat_df['Enrichr gene page'] = [\"https://maayanlab.cloud/Enrichr/#find!gene=%s\"%i for i in stat_df.index]\n",
    "        filename = \"pert_stat_%s_%s.tsv\"%(lib.replace(\" \",\"_\"), direction)\n",
    "        stat_df.to_csv(filename, sep=\"\\t\")\n",
    "        stat_df['Enrichr gene page'] = stat_df['Enrichr gene page'].apply(make_clickable)\n",
    "        stat_html = stat_df.head(25).to_html(escape=False)\n",
    "        display(HTML(stat_html))\n",
    "    else:\n",
    "        stat_df = pd.DataFrame(0, index=index, columns=[\"count\", \"z-score\", \"p-value\"])\n",
    "        stat_df['count'] = filtered_df.sum(1)\n",
    "        # Compute zstat and p value\n",
    "        stat_df[\"z-score\"] = (filtered_df.mean(1) - empirical_stat[\"mean\"]) / empirical_stat[\"std\"]\n",
    "        stat_df[\"p-value\"] = stat_df['z-score'].apply(lambda x: 1-st.norm.cdf(x))\n",
    "\n",
    "        if consensus_method == 'zscore':\n",
    "            #Filter by p-value\n",
    "            stat_df = stat_df[stat_df[\"p-value\"]<alpha].sort_values(by=[\"z-score\"], ascending=False)\n",
    "        else:\n",
    "            stat_df = stat_df.sort_values(by=[\"count\"], ascending=False)\n",
    "\n",
    "        filename = \"pert_stat_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "        stat_df.to_csv(filename, sep=\"\\t\")\n",
    "        display(stat_df.head(25))\n",
    "    display(Markdown(\"**Table %d** Top 25 mimicker perturbagens using %s dataset ([download](./%s))\"%\n",
    "                     (table, lib, filename)))\n",
    "\n",
    "    table+=1\n",
    "\n",
    "    consensus = df.loc[stat_df.index[0:top_perts]]\n",
    "    consensus_norm = quantile_normalize(consensus)\n",
    "#     consensus_norm = consensus.subtract(empirical_stat.loc[consensus.index, \"mean\"], axis=0).divide(empirical_stat.loc[consensus.index, \"std\"], axis=0)\n",
    "    if len(consensus.index) > top_perts:\n",
    "        consensus = consensus.loc[consensus.index[:top_perts]]\n",
    "    filename = \"consensus_matrix_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "    consensus.to_csv(filename, sep=\"\\t\")\n",
    "    display(consensus.head())\n",
    "    display(Markdown(\"**Table %d** Consensus mimicker perturbation matrix for %s ([download](./%s))\"%\n",
    "                     (table, lib, filename)))\n",
    "\n",
    "    table+=1\n",
    "    \n",
    "    plot_label = \"%s %s\"%(lib, direction)\n",
    "    figure = get_tsne(consensus_norm, plot_label, figure)\n",
    "    figure = get_umap(consensus_norm, plot_label, figure)\n",
    "    \n",
    "\n",
    "    label = \"Clustergrammer of consensus mimicker perturbation of %s (quantile normalized scores)\"%lib\n",
    "    name = \"clustergrammer_%s_%s.tsv\"%(library.replace(\" \", \"_\"), direction)\n",
    "    figure = clustergrammer(consensus_norm, name, figure, label)\n",
    "\n",
    "    label = \"Heatmap of consensus mimicker perturbation of %s (quantile normalized scores)\"%lib\n",
    "    name = \"heatmap_%s_%s.png\"%(library.replace(\" \", \"_\"), direction)\n",
    "    figure = heatmap(consensus_norm, name, figure, label)\n",
    "    \n",
    "    if lib in drug_page:\n",
    "        display(Markdown(\"### Drugmonizome enrichment analysis for the consensus drugs\"))\n",
    "        for d in drugmonizome_datasets:\n",
    "            label = \"Drugmonizome %s hits for %s %s\"%(d, lib, direction)\n",
    "            figure = get_drugmonizome_plot(consensus, label, figure, d)\n",
    "    elif lib in gene_page:\n",
    "        display(Markdown(\"### Enrichr terms for the consensus genes\"))\n",
    "        label = \"%s %s\"%(lib, direction)\n",
    "        figure = enrichment(consensus, label, figure)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reversers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "direction = \"reversers\"\n",
    "for lib in datasets:\n",
    "    library = dataset_map[lib]\n",
    "    display(Markdown(\"### %s\"%lib))\n",
    "    index = set()\n",
    "    pert_dict = enriched[direction][lib]\n",
    "    for v in pert_dict.values():\n",
    "        index = index.union(v.keys())\n",
    "    df = pd.DataFrame(0, index=index, columns=pert_dict.keys())\n",
    "    for k,v in pert_dict.items():\n",
    "        for pert, cnt in v.items():\n",
    "            df.at[pert, k] += cnt\n",
    "    df = df.loc[df.sum(1).sort_values(ascending=False).index]\n",
    "    filename = \"pert_matrix_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "    df.to_csv(filename, sep=\"\\t\")\n",
    "#     display(df.head())\n",
    "#     display(Markdown(\"**Table %d** Reverser perturbagens using %s dataset ([download](./%s))\"%\n",
    "#                      (table, lib, filename)))\n",
    "    table += 1\n",
    "    # stat_df = pd.DataFrame(0, index=df.index, columns=[\"count\", \"z-score\", \"p-value\"])\n",
    "    empirical_stat = pd.read_csv(S3_PREFIX + \"%s_%s.tsv\" % (library, direction), sep=\"\\t\", index_col=0)\n",
    "    df = df[(df>0).sum(1) > len(df.columns) * perc]\n",
    "    index = set(df.index).intersection(empirical_stat.index)\n",
    "    filtered_df = df.loc[index]    \n",
    "    if lib in gene_page:\n",
    "        stat_df = pd.DataFrame(0, index=index, columns=[\"count\", \"z-score\", \"p-value\", \"Enrichr gene page\"])\n",
    "        stat_df['count'] = filtered_df.sum(1)\n",
    "        # Compute zstat and p value\n",
    "        stat_df[\"z-score\"] = (filtered_df.mean(1) - empirical_stat[\"mean\"]) / empirical_stat[\"std\"]\n",
    "        stat_df[\"p-value\"] = stat_df['z-score'].apply(lambda x: 1-st.norm.cdf(x))\n",
    "\n",
    "        if consensus_method == 'zscore':\n",
    "            #Filter by p-value\n",
    "            stat_df = stat_df[stat_df[\"p-value\"]<alpha].sort_values(by=[\"z-score\"], ascending=False)\n",
    "        else:\n",
    "            stat_df = stat_df.sort_values(by=[\"count\"], ascending=False)\n",
    "\n",
    "        stat_df['Enrichr gene page'] = [\"https://maayanlab.cloud/Enrichr/#find!gene=%s\"%i for i in stat_df.index]\n",
    "        filename = \"pert_stat_%s_%s.tsv\"%(lib.replace(\" \",\"_\"), direction)\n",
    "        stat_df.to_csv(filename, sep=\"\\t\")\n",
    "        stat_df['Enrichr gene page'] = stat_df['Enrichr gene page'].apply(make_clickable)\n",
    "        stat_html = stat_df.head(25).to_html(escape=False)\n",
    "        display(HTML(stat_html))\n",
    "    else:\n",
    "        stat_df = pd.DataFrame(0, index=index, columns=[\"count\", \"z-score\", \"p-value\"])\n",
    "        stat_df['count'] = filtered_df.sum(1)\n",
    "        # Compute zstat and p value\n",
    "        stat_df[\"z-score\"] = (filtered_df.mean(1) - empirical_stat[\"mean\"]) / empirical_stat[\"std\"]\n",
    "        stat_df[\"p-value\"] = stat_df['z-score'].apply(lambda x: 1-st.norm.cdf(x))\n",
    "\n",
    "        if consensus_method == 'zscore':\n",
    "            #Filter by p-value\n",
    "            stat_df = stat_df[stat_df[\"p-value\"]<alpha].sort_values(by=[\"z-score\"], ascending=False)\n",
    "        else:\n",
    "            stat_df = stat_df.sort_values(by=[\"count\"], ascending=False)\n",
    "\n",
    "        filename = \"pert_stat_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "        stat_df.to_csv(filename, sep=\"\\t\")\n",
    "        display(stat_df.head(25))\n",
    "    display(Markdown(\"**Table %d** Top 25 reverser perturbagens using %s dataset ([download](./%s))\"%\n",
    "                     (table, lib, filename)))\n",
    "\n",
    "    table+=1\n",
    "\n",
    "    consensus = df.loc[stat_df.index[0:top_perts]]\n",
    "    consensus_norm = quantile_normalize(consensus)\n",
    "#     consensus_norm = consensus.subtract(empirical_stat.loc[consensus.index, \"mean\"], axis=0).divide(empirical_stat.loc[consensus.index, \"std\"], axis=0)\n",
    "    if len(consensus.index) > top_perts:\n",
    "        consensus = consensus.loc[consensus.index[:top_perts]]\n",
    "    filename = \"consensus_matrix_%s_%s.tsv\"%(library.replace(\" \",\"_\"), direction)\n",
    "    consensus.to_csv(filename, sep=\"\\t\")\n",
    "    display(consensus.head())\n",
    "    display(Markdown(\"**Table %d** Consensus reverser perturbation matrix for %s ([download](./%s))\"%\n",
    "                     (table, lib, filename)))\n",
    "\n",
    "    table+=1\n",
    "    \n",
    "    plot_label = \"%s %s\"%(lib, direction)\n",
    "    figure = get_tsne(consensus_norm, plot_label, figure)\n",
    "    figure = get_umap(consensus_norm, plot_label, figure)\n",
    "    \n",
    "\n",
    "    label = \"Clustergrammer of consensus reverser perturbation of %s (quantile normalized scores)\"%lib\n",
    "    name = \"clustergrammer_%s_%s.tsv\"%(library.replace(\" \", \"_\"), direction)\n",
    "    figure = clustergrammer(consensus_norm, name, figure, label)\n",
    "\n",
    "    label = \"Heatmap of consensus reverser perturbation of %s (quantile normalized scores)\"%lib\n",
    "    name = \"heatmap_%s_%s.png\"%(library.replace(\" \", \"_\"), direction)\n",
    "    figure = heatmap(consensus_norm, name, figure, label)\n",
    "    \n",
    "    \n",
    "    if lib in drug_page:\n",
    "        display(Markdown(\"### Drugmonizome enrichment analysis for the consensus drugs\"))\n",
    "        for d in drugmonizome_datasets:\n",
    "            label = \"Drugmonizome %s hits for %s %s\"%(d, lib, direction)\n",
    "            figure = get_drugmonizome_plot(consensus, label, figure, d)\n",
    "    elif lib in gene_page:\n",
    "        display(Markdown(\"### Enrichr terms for the consensus genes\"))\n",
    "        label = \"%s %s\"%(lib, direction)\n",
    "        figure = enrichment(consensus, label, figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter-default",
   "language": "python",
   "name": "jupyter-default"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
